
<!DOCTYPE html>


<html lang="en" data-content_root="../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>10.1: Data Preparation and Feature Engineering &#8212; 2026 Cheminformatics OLCC</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../../../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'content/modules/10_SupervisedML/10_1_data_prep';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Appendix 1.1: Getting Set Up" href="../../appendices/App_1/README.html" />
    <link rel="prev" title="10.0: Intro" href="10_0_Intro.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../intro.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/logo.png" class="logo__image only-light" alt="2026 Cheminformatics OLCC - Home"/>
    <script>document.write(`<img src="../../../_static/logo.png" class="logo__image only-dark" alt="2026 Cheminformatics OLCC - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../intro.html">
                    Welcome to your Jupyter Book
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Core Modules</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../01_PythonPrimer/README.html">Module 1 Python Primers</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../01_PythonPrimer/01_1_python-basics.html">1.1 Python Basics</a></li>


<li class="toctree-l2"><a class="reference internal" href="../01_PythonPrimer/01_2_python-intermediate.html">1.2 Intermediate Python</a></li>


</ul>
</details></li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="README.html">Module 10: Supervised Machine Learning</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="10_0_Intro.html">10.0: Intro</a></li>




<li class="toctree-l2 current active"><a class="current reference internal" href="#">10.1: Data Preparation and Feature Engineering</a></li>






</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Appendices</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../appendices/App_1/README.html">Appendix 1.1: Getting Set Up</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../appendices/App_1/A_01-1_OS.html">Appendix 1: Operating System</a></li>



</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../appendices/App_10/README.html">Appendix 10: Supervised Machine Learning</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../appendices/App_10/A_10-1AID_Selector.html">A10.1 BioAssay Screening: Enough Actives</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../appendices/App_10/A_10-2NaiveBayes.html">A10.2 Bayes Theorem as an Inference</a></li>



</ul>
</details></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-launch-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Launch interactive content">
    <i class="fas fa-rocket"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://nanohub.org/tools/datachemolcc/hub/user-redirect/git-pull?repo=https%3A//github.com/rebelford/2026Cheminformatics&urlpath=lab/tree/2026Cheminformatics/content/modules/10_SupervisedML/10_1_data_prep.ipynb&branch=main" target="_blank"
   class="btn btn-sm dropdown-item"
   title="Launch on JupyterHub"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  
    <img alt="JupyterHub logo" src="../../../_static/images/logo_jupyterhub.svg">
  </span>
<span class="btn__text-container">JupyterHub</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/rebelford/2026Cheminformatics" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/rebelford/2026Cheminformatics/issues/new?title=Issue%20on%20page%20%2Fcontent/modules/10_SupervisedML/10_1_data_prep.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../../_sources/content/modules/10_SupervisedML/10_1_data_prep.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>10.1: Data Preparation and Feature Engineering</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">10.1: Data Preparation and Feature Engineering</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#import-bioactivity-data-from-pubchem">1. Import bioactivity data from PubChem</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#directory-architecture">1.1 Directory Architecture</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#helper-function-make-fixed-filename">1.2 Helper Function: make_fixed_filename</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#data-cleaning">2  Data Cleaning</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#inspect-data-and-remove-first-3-rows">2.1 Inspect data and remove first 3 rows</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#save-versioned-dataframe-as-csv">2.2 Save Versioned DataFrame as CSV</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#helper-function-make-filename">2.2.1 Helper Function: make_filename</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reload-saved-file-into-dataframe">2.2.2 Reload saved file into dataframe</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#inspect-column-names">2.3 Inspect column names</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#clean-column-names">2.3.1 Clean column names</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#inspect-activity-outcomes">3. Inspect Activity Outcomes</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#inspect-activity-summaries">3.1 Inspect Activity summaries</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#select-active-inactive-compounds-for-model-building">4. Select active/inactive compounds for model building</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#select-active-and-inactive-compounds">4.1 Select active and inactive compounds</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#drop-substances-without-associated-cids">4.2 Drop substances without associated CIDs.</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#remove-cids-with-conflicting-activities">4.3 Remove CIDs with conflicting activities</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#remove-redundant-data">4.4 Remove redundant data</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#save-as-aid743139-clean-cid">Save as AID743139_clean_CID…</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#adding-numeric-activity-classes">4.5 Adding “numeric” activity classes</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#create-a-smaller-data-frame-that-only-contains-cids-and-activities">4.6 Create a smaller data frame that only contains CIDs and activities.</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#save-as-aid743139-activity-cid">4.6.1 Save as AID743139_activity_CID</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#acquire-standardized-pubchem-smiles">5 Acquire standardized PubChem SMILES</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#deal-with-salts-and-mixtures">5.1  Deal with salts and mixtures</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#comparing-structures-using-inchi">5.2 Comparing structures using InChI</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#save-as-aid743139-nosalt-pcsmiles">Save as AID743139_noSalt_pcSMILES…</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#convert-smiles-to-binary-data-for-model-input">6 Convert SMILES to Binary Data for Model Input</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-maccs-keys-from-smiles">6.1 Generate MACCS keys from SMILES</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#save-as-features-aid743139-maccs-nosalt">Save as features/AID743139_MACCS_noSalt…</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#merge-activity-data-and-fingerprint-information">6.2 Merge activity data and fingerprint information</a></li>
</ul>
</li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="data-preparation-and-feature-engineering">
<h1>10.1: Data Preparation and Feature Engineering<a class="headerlink" href="#data-preparation-and-feature-engineering" title="Link to this heading">#</a></h1>
<div class="alert alert-block alert-warning">
<h1>Learning Objectives</h1>
<p><strong>Purpose</strong>  Introduce students to the data preparation steps needed to create a dataset suitable for machine learning.  Introduce students to feature engineering and the generation of binary fingerprints from SMILES strings.</p>
<p><strong>Students Learn</strong></p>
<ul class="simple">
<li><p>How to obtain BioAssay data from PubChem</p></li>
<li><p>How to curate the data</p></li>
<li><p>How to save data and maintain the data’s provenance</p></li>
</ul>
<p><strong>Core Activities</strong></p>
<ul class="simple">
<li><p>Load PubChem Data</p></li>
<li><p>Curate compounds (salts, largest fragments, conflicts…)</p></li>
<li><p>Generate fingerprints</p></li>
<li><p>Save compounds and metadata</p></li>
</ul>
<p><strong>Prior Knowledge</strong></p>
<ul class="simple">
<li><p>Programmatic access to PubChem</p></li>
<li><p>Familiarity with SMILES and InChI</p></li>
<li><p>Familiarity with RDKit fingerprint generation</p></li>
<li><p>Familiarity with Pandas</p></li>
</ul>
<p><strong>Reading links</strong></p>
<ul class="simple">
<li><p><a class="reference external" href="https://rebelford.github.io/py4scibook/content/module3-packages1/03-2-pandas.html">Pandas Chapter of Py4Sci</a></p></li>
</ul>
</div></section>
<section class="tex2jax_ignore mathjax_ignore" id="import-bioactivity-data-from-pubchem">
<h1>1. Import bioactivity data from PubChem<a class="headerlink" href="#import-bioactivity-data-from-pubchem" title="Link to this heading">#</a></h1>
<p>In this notebook, we will develop a prediction model for small molecule’s activity against human aromatase (https://pubchem.ncbi.nlm.nih.gov/protein/EAW77416), which is encoded by the CYP19A1 gene (https://pubchem.ncbi.nlm.nih.gov/gene/1588). The model will predict the activity of a molecule based on the structure of the molecule (represented with molecular fingerprints).</p>
<p>Endocrine disrupting chemicals (EDCs) interfere with the biosynthesis and normal functions of steroid hormones including estrogen and androgen in the body. Aromatase catalyzes the conversion of androgen to estrogen and plays a key role in maintaining the androgen and estrogen balance in many of the EDC-sensitive organs.</p>
<p>For model development, we will use the Tox21 bioassay data for human aromatase, archived in PubChem (https://pubchem.ncbi.nlm.nih.gov/bioassay/743139).  The bioactivity data presented on this page can be downloaded by clicking the “Download” button available on this page and then read the data into a data frame.  Alternatively, you can directly load the data into a data frame as shown in the cell below.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>

<span class="n">url</span> <span class="o">=</span> <span class="s1">&#39;https://pubchem.ncbi.nlm.nih.gov/assay/pcget.cgi?query=download&amp;record_type=datatable&amp;actvty=all&amp;response_type=save&amp;aid=743139&#39;</span>
<span class="n">df_raw</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_raw</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(10489, 16)
</pre></div>
</div>
</div>
</div>
<section id="directory-architecture">
<h2>1.1 Directory Architecture<a class="headerlink" href="#directory-architecture" title="Link to this heading">#</a></h2>
<p>In this activity we are going to generate multiple fingerprints and eventually run different ML programs on different BioAssays. Below is a sample hierarchy structure for each bioassay, which is this class will be in the downloads folder of your module working directory.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>10_SupervisedML/
├── 10_introduction.ipynb
│── 10_1_data_prep.ipynb
│── 10_2_NB_model_construction_workflow.ipynb
│── 10_3_pipelines_and_inference.ipynb
│── 10_4_model_eval_interpretation.ipynb
│── 10_5_model_comparison_exp_design.ipynb
├── data/                      ← algorithm-agnostic artifacts
│   └── AID743139/
│       ├── raw/               ← raw PubChem exports
│       │   └── AID743139_pubchem_raw_20251224.csv
│       ├── curated/           ← chemically curated datasets
│       │   └── AID743139_Activity_CID_20260111_v1.csv
│       ├── features/          ← feature-level artifacts (post-representation)
│       │   └── AID743139_MACCS_activities_noSalt_20260104_v1.csv
│       │   ├── maccs_variance_mask_20260104_v1.npy
│       │   └── feature_metadata_20260104_v1.json
│       └── splits/            ← experimental splits
│           └── 90_10/
│               ├── arrays/
│               │   ├── X_train.npy
│               │   ├── X_test.npy
│               │   ├── y_train.npy
│               │   └── y_test.npy
│               └── split_metadata.json
│
├── models/                    ← algorithm-specific artifacts
│   └── AID743139/
│       ├── nb_maccs_20260109_v1.joblib
│       ├── dt_maccs_20260110_v1.joblib
│       └── rf_maccs_20260112_v1.joblib
│
├── results/                   ← evaluation outputs
│   └── AID743139/
│       ├── nb/
│       ├── dt/
│       └── comparison_tables/
│
</pre></div>
</div>
</section>
<section id="helper-function-make-fixed-filename">
<h2>1.2 Helper Function: make_fixed_filename<a class="headerlink" href="#helper-function-make-fixed-filename" title="Link to this heading">#</a></h2>
<p>In the following code cell we are going to use the helper function <code class="docutils literal notranslate"><span class="pre">make_fixed_filename</span></code> to save this data in the raw data folder of the directory structure above.  These helper functions are provided in the cinf26pk package we have been building in this class, and the code is provided in the green box below.  If you do not have access to the package you can substitute it’s code for the import statement above</p>
<p>Note, we will be using two helper functions for saving files and both will generate the directory architecture based on your current working directory:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">make_fixed_filename</span></code> - for deterministic filenames with date stamp, this allows us to identify if the data file at PubChem as changed in the future, but does not rewrite a new version if we redownload it multiple times on the same day.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">make_filename</span></code> - for versioned filenames with date stamp, this allows us to save our curated data and features with version numbers so we can experiment and not lose our work.</p></li>
</ul>
<div class="alert alert-block alert-success">
<strong>Helper Function: make_fixed_filename</strong>
<p>The following file was imported from the class package and used in the preceeding script</p>
<div style="
    background-color: #efffff;
    color: #000000;
    padding: 10px;
    border-radius: 4px;
    border: 1px solid #dddddd;
    margin-top: 10px;
  ">
<details>
    <summary>Code</summary>
 <p>The `make_fixed_filename` helpeer function was created in the class package pvhelpfct.py.  It is for deterministic in contrast to versioning names, but this does at a date stamp to the filename.  We use this helper function for data sets we download. After we work them up we use the `make_filename` helper function that versions the filename.  Below is the code for the function, although we will import it from the class package in this example</p>
<pre><code>def make_fixed_filename(
    stem: str,
    ext: str = "csv",
    add_date: bool = True
) -> str:
    """
    Create a deterministic filename of the form:
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>    stem[_YYYYMMDD].ext

Intended for canonical or authoritative data artifacts
that should not auto-version.
&quot;&quot;&quot;
if add_date:
    date_str = datetime.now().strftime(&quot;%Y%m%d&quot;)
    return f&quot;{stem}_{date_str}.{ext}&quot;
else:
    return f&quot;{stem}.{ext}&quot;
</pre></div>
</div>
<p></code></pre></p>
</details>
</div>
</div>
<p><strong>Note: For the following code cell to generate the above directory structure your notebook must be in the 10_SupervisedML/notebooks folder.</strong> This code will create the <code class="docutils literal notranslate"><span class="pre">/data/AID743139/raw</span></code> directory path if it does not exist.  If you are working with a different bioassay you need to change the AID in the path to the new one.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">cinf26pk.core</span><span class="w"> </span><span class="kn">import</span> <span class="n">filenames</span>

<span class="n">PROJECT_ROOT</span><span class="o">=</span><span class="n">Path</span><span class="o">.</span><span class="n">cwd</span><span class="p">()</span>
<span class="n">RAW</span> <span class="o">=</span> <span class="n">PROJECT_ROOT</span><span class="o">/</span><span class="s2">&quot;data&quot;</span><span class="o">/</span> <span class="s2">&quot;AID743139&quot;</span> <span class="o">/</span> <span class="s2">&quot;raw&quot;</span>
<span class="n">RAW</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">fname</span> <span class="o">=</span> <span class="n">filenames</span><span class="o">.</span><span class="n">make_fixed_filename</span><span class="p">(</span>
    <span class="n">stem</span><span class="o">=</span><span class="s2">&quot;AID743139_pubchem_raw&quot;</span><span class="p">,</span>
    <span class="n">ext</span><span class="o">=</span><span class="s2">&quot;csv&quot;</span><span class="p">,</span>
    <span class="n">add_date</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="n">df_raw</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">RAW</span> <span class="o">/</span> <span class="n">fname</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="alert alert-block alert-success">
<strong>Python Check</strong>
<p>
The above code saves the raw PubChem dataset to a well-defined location
using a consistent naming policy. 
    <center><strong>Explain the code.</strong></center>
</p>
  <div style="
    background-color: #efffff;
    color: #000000;
    padding: 10px;
    border-radius: 4px;
    border: 1px solid #dddddd;
    margin-top: 10px;
  ">
<details>
  <summary>Answer</summary>
  <p>
  There are two things going on here.
  <ul>
      <li>We are creating a path</li>
      <li>We are dating our filenames</li>
  </ul>
  The first thing we need to do is understand the following script
  </p>
  <pre><code>
PROJECT_ROOT = Path.cwd() # protject root (10_SupervisedML)
RAW = PROJECT_ROOT / "data" / "AID743139" / "raw"
RAW.mkdir(parents=True, exist_ok=True)
  </code></pre>
<p>
<ul>
  <li>
    The first line defines the project root as the parent of the current working directory.
    Since this notebook is run from the <code>notebooks/</code> directory, its parent
    directory (<code>10_SupervisedML</code>) becomes the project root. We capitalize
    <code>PROJECT_ROOT</code> because it represents an invariant configuration value.
  </li>
  <li>
    The second line constructs the path from the project root to the directory where
    the raw data will be stored.
  </li>
  <li>
    The third line creates the directory (and any missing parent directories) if they
    do not already exist, making the code safe to rerun.
  </li>
</ul>
 </p>
  <p>
  The helper function <code>make_fixed_filename</code> is then used to construct
  the output filename:
  </p>
 Create deterministic filename with date stamp
  <pre><code>
from cinf26.core import filenames 
fname = filenames.make_fixed_filename(
    stem="AID743139_pubchem_raw",
    ext="csv",
    add_date=True
)
<p></code></pre></p>
<p>This uses the helper function `make_fixed_filename`, which is described in the next dropdown box</p>
  <p>
  Finally, the DataFrame is written to disk:
  </p>
  <pre><code>
df_raw.to_csv(RAW / fname, index=False)
  </code></pre>
  <ul>
    <li><strong>Use named path objects</strong> to represent meaningful locations in the workflow, rather than embedding directory structures inside string literals.</li>
    <li><strong>Use helper functions</strong> to enforce consistent naming policies for saved artifacts, so filenames encode intent rather than ad-hoc decisions.</li>
  </ul>
</details>
</div> 
</div>
<ol class="arabic simple">
<li><p><strong>Restart Kernel and Load the code into a dataframe called df_raw</strong>  (see green box below for help)</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Write Code here </span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>

<span class="n">PROJECT_ROOT</span><span class="o">=</span><span class="n">Path</span><span class="o">.</span><span class="n">cwd</span><span class="p">()</span>
<span class="n">RAW</span> <span class="o">=</span> <span class="n">PROJECT_ROOT</span><span class="o">/</span><span class="s2">&quot;data&quot;</span><span class="o">/</span> <span class="s2">&quot;AID743139&quot;</span> <span class="o">/</span> <span class="s2">&quot;raw&quot;</span>
<span class="n">RAW</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">df_raw</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">RAW</span> <span class="o">/</span> <span class="s2">&quot;AID743139_pubchem_raw_20260117.csv&quot;</span><span class="p">)</span>
<span class="n">df_raw</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">7</span><span class="p">)</span>  <span class="c1"># gives the first 7 rows of the dataframe</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>PUBCHEM_RESULT_TAG</th>
      <th>PUBCHEM_SID</th>
      <th>PUBCHEM_CID</th>
      <th>PUBCHEM_EXT_DATASOURCE_SMILES</th>
      <th>PUBCHEM_ACTIVITY_OUTCOME</th>
      <th>PUBCHEM_ACTIVITY_SCORE</th>
      <th>PUBCHEM_ACTIVITY_URL</th>
      <th>PUBCHEM_ASSAYDATA_COMMENT</th>
      <th>Activity Summary</th>
      <th>Antagonist Activity</th>
      <th>Antagonist Potency (uM)</th>
      <th>Antagonist Efficacy (%)</th>
      <th>Viability Activity</th>
      <th>Viability Potency (uM)</th>
      <th>Viability Efficacy (%)</th>
      <th>Sample Source</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>RESULT_TYPE</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>STRING</td>
      <td>STRING</td>
      <td>FLOAT</td>
      <td>FLOAT</td>
      <td>STRING</td>
      <td>FLOAT</td>
      <td>FLOAT</td>
      <td>STRING</td>
    </tr>
    <tr>
      <th>1</th>
      <td>RESULT_DESCR</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>Type of compound activity based on both the ar...</td>
      <td>Type of compound activity in the aromatase ant...</td>
      <td>The concentration of sample yielding half-maxi...</td>
      <td>Percent inhibition of aromatase.</td>
      <td>Type of compound activity in the cell viabilit...</td>
      <td>The concentration of sample yielding half-maxi...</td>
      <td>Percent inhibition of cell viability.</td>
      <td>Where sample was obtained.</td>
    </tr>
    <tr>
      <th>2</th>
      <td>RESULT_UNIT</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>MICROMOLAR</td>
      <td>PERCENT</td>
      <td>NaN</td>
      <td>MICROMOLAR</td>
      <td>PERCENT</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3</th>
      <td>1</td>
      <td>144203552.0</td>
      <td>12850184.0</td>
      <td>C(C(=O)[C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O.C(...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>4</th>
      <td>2</td>
      <td>144203553.0</td>
      <td>89753.0</td>
      <td>C([C@H]([C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O)O...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>5</th>
      <td>3</td>
      <td>144203554.0</td>
      <td>9403.0</td>
      <td>C[C@]12CC[C@H]3[C@H]([C@@H]1CC[C@@H]2OC(=O)CCC...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>6</th>
      <td>4</td>
      <td>144203555.0</td>
      <td>13218779.0</td>
      <td>C[C@@]12CC[C@@H](C1(C)C)C[C@H]2OC(=O)CSC#N</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>NCI</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="alert alert-block alert-success">
<strong>Pandas Check</strong>
<p>How do you open the raw csv file in Pandas?</p>
  <div style="
    background-color: #efffff;
    color: #000000;
    padding: 10px;
    border-radius: 4px;
    border: 1px solid #dddddd;
    margin-top: 10px;
  ">
<details>
    <summary>Answer</summary>
There are two approaches, use the predefined path variable, or write the entire path as a string. For the first to work you must define the path, which was done in a preceding cell.  Of course you must use the filename of the file you saved. 
<pre><code>
import pandas as pd
df_raw = pd.read_csv(RAW / "AID743139_pubchem_raw_20251230.csv")
#df_raw = pd.read_csv("../data/AID743139/raw/AID743139_pubchem_raw_20251230.csv")  
</code></pre>
Note, if you hard code the path as a string, you need to start with "../" and move up one level (`10_SupervisedML`). That is  the notebook is being run in the `10/ML/notebooks` directory and you want your data in the `10_SupervisedML/data` directory. 
</div>
</div>
<ol class="arabic simple" start="2">
<li><p><strong>Display the first 5 rows of the df_raw</strong></p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Write Code here </span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="n">df_raw</span><span class="o">.</span><span class="n">head</span><span class="p">()</span> 
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>PUBCHEM_RESULT_TAG</th>
      <th>PUBCHEM_SID</th>
      <th>PUBCHEM_CID</th>
      <th>PUBCHEM_EXT_DATASOURCE_SMILES</th>
      <th>PUBCHEM_ACTIVITY_OUTCOME</th>
      <th>PUBCHEM_ACTIVITY_SCORE</th>
      <th>PUBCHEM_ACTIVITY_URL</th>
      <th>PUBCHEM_ASSAYDATA_COMMENT</th>
      <th>Activity Summary</th>
      <th>Antagonist Activity</th>
      <th>Antagonist Potency (uM)</th>
      <th>Antagonist Efficacy (%)</th>
      <th>Viability Activity</th>
      <th>Viability Potency (uM)</th>
      <th>Viability Efficacy (%)</th>
      <th>Sample Source</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>RESULT_TYPE</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>STRING</td>
      <td>STRING</td>
      <td>FLOAT</td>
      <td>FLOAT</td>
      <td>STRING</td>
      <td>FLOAT</td>
      <td>FLOAT</td>
      <td>STRING</td>
    </tr>
    <tr>
      <th>1</th>
      <td>RESULT_DESCR</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>Type of compound activity based on both the ar...</td>
      <td>Type of compound activity in the aromatase ant...</td>
      <td>The concentration of sample yielding half-maxi...</td>
      <td>Percent inhibition of aromatase.</td>
      <td>Type of compound activity in the cell viabilit...</td>
      <td>The concentration of sample yielding half-maxi...</td>
      <td>Percent inhibition of cell viability.</td>
      <td>Where sample was obtained.</td>
    </tr>
    <tr>
      <th>2</th>
      <td>RESULT_UNIT</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>MICROMOLAR</td>
      <td>PERCENT</td>
      <td>NaN</td>
      <td>MICROMOLAR</td>
      <td>PERCENT</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3</th>
      <td>1</td>
      <td>144203552.0</td>
      <td>12850184.0</td>
      <td>C(C(=O)[C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O.C(...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>4</th>
      <td>2</td>
      <td>144203553.0</td>
      <td>89753.0</td>
      <td>C([C@H]([C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O)O...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>NCI</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="alert alert-block alert-success">
<strong>Pandas Check</strong>
<p>Why are the CID and SID columns floats and not integers</p>
  <div style="
    background-color: #efffff;
    color: #000000;
    padding: 10px;
    border-radius: 4px;
    border: 1px solid #dddddd;
    margin-top: 10px;
  ">
<details>
    <summary>Answer</summary>
<p>Pandas infers column data types when a CSV file is read and even though CIDs and SIDs are integers, the first few rows contain missing values (there is no CID or SID) and so Pandas converts them to NaN (Not a Number), which is actually a IEEE standard and treated as a float, allowing Pandas to handle missing number. As a result, Pandas promotes the column to <code>float64</code> in order to accommodate <code>NaN</code>. If all values were integers then the column would be integers. Once those rows are removed, the columns can safely be converted to integers using the <code>.astype(int)</code> method. This conversion will fail if any values are still missing (<code class="docutils literal notranslate"><span class="pre">NaN</span></code>). For more information see section:   <strong><a href="https://rebelford.github.io/py4scibook/content/module3-packages1/03-2-pandas.html#handling-missing-data">Section 4.3 Handling Missing Data in Py4Sci Pandas Chapter</a></strong></p>
<p>Note that a column in a DataFrame is actually a Pandas <code>Series</code>, and the column label is the
Series name. A DataFrame can be viewed as a table of Series objects that all share a common row index.
Therefore, the following operation is calling a <strong>Series method</strong> on a column, and this could convert the values to integers, but will not work if one of the values is NaN.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;column-label&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;column-label&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
</pre></div>
</div>
<p>For more information see section:   <strong><a href="https://rebelford.github.io/py4scibook/content/module3-packages1/03-2-pandas.html#handling-missing-data">Section 4.3 Handling Missing Data in Py4Sci Pandas Chapter</a></strong></p>
</details>
</div>
</div></section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="data-cleaning">
<h1>2  Data Cleaning<a class="headerlink" href="#data-cleaning" title="Link to this heading">#</a></h1>
<p>Data cleaning is a critical step in our workflow that ensures the quality, consistency, and reliability of a dataset before we build a model. Raw data often contains missing values, duplicates, inconsistent formatting, or mislabeled entries that can lead to incorrect conclusions or poorly performing models. By identifying and resolving these issues at the start of our model design, we create a dataset that accurately reflects the information needed for meaningful interpretation and predictive modeling.</p>
<section id="inspect-data-and-remove-first-3-rows">
<h2>2.1 Inspect data and remove first 3 rows<a class="headerlink" href="#inspect-data-and-remove-first-3-rows" title="Link to this heading">#</a></h2>
<p>Looking at our imported data, lines 0-2 provide the descriptions for each column (data type, descriptions, units, etc).  These rows need be removed. We will do so with a <strong>slice operation</strong> to select only those rows that contain the data we will use in the model (See <a class="reference external" href="https://rebelford.github.io/py4scibook/content/module3-packages1/03-2-pandas.html#using-iloc-on-a-dataframe">Py4Sci: using .iloc on a dataframe</a>.)</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_curated</span> <span class="o">=</span> <span class="n">df_raw</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">3</span><span class="p">:]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>  <span class="c1">#the [3:] indicates a slice opperation in pandas dataframes. </span>
<span class="c1"># It selects all rows starting at position 3 and continues to the end, we than save</span>
<span class="c1">#over  the first dataframe, effectively removing tows 0,1 and2</span>
<span class="n">df_curated</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>PUBCHEM_RESULT_TAG</th>
      <th>PUBCHEM_SID</th>
      <th>PUBCHEM_CID</th>
      <th>PUBCHEM_EXT_DATASOURCE_SMILES</th>
      <th>PUBCHEM_ACTIVITY_OUTCOME</th>
      <th>PUBCHEM_ACTIVITY_SCORE</th>
      <th>PUBCHEM_ACTIVITY_URL</th>
      <th>PUBCHEM_ASSAYDATA_COMMENT</th>
      <th>Activity Summary</th>
      <th>Antagonist Activity</th>
      <th>Antagonist Potency (uM)</th>
      <th>Antagonist Efficacy (%)</th>
      <th>Viability Activity</th>
      <th>Viability Potency (uM)</th>
      <th>Viability Efficacy (%)</th>
      <th>Sample Source</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>3</th>
      <td>1</td>
      <td>144203552.0</td>
      <td>12850184.0</td>
      <td>C(C(=O)[C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O.C(...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>4</th>
      <td>2</td>
      <td>144203553.0</td>
      <td>89753.0</td>
      <td>C([C@H]([C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O)O...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>5</th>
      <td>3</td>
      <td>144203554.0</td>
      <td>9403.0</td>
      <td>C[C@]12CC[C@H]3[C@H]([C@@H]1CC[C@@H]2OC(=O)CCC...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>6</th>
      <td>4</td>
      <td>144203555.0</td>
      <td>13218779.0</td>
      <td>C[C@@]12CC[C@@H](C1(C)C)C[C@H]2OC(=O)CSC#N</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>7</th>
      <td>5</td>
      <td>144203556.0</td>
      <td>142766.0</td>
      <td>C1=CC=C(C(=C1)C(=O)O)O.C1=CC2=C(C(=C1)O)N=CC=C2</td>
      <td>Inconclusive</td>
      <td>25.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inconclusive antagonist (cytotoxic)</td>
      <td>active antagonist</td>
      <td>15.5454</td>
      <td>-115.803</td>
      <td>active antagonist</td>
      <td>14.9601</td>
      <td>-76.8218</td>
      <td>NCI</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Now that we have removed the first three rows we can convert the CIDs and SIDs to integers using Pandas nullable <code class="docutils literal notranslate"><span class="pre">&quot;Int64&quot;</span></code> type that allows NaN (this is not <code class="docutils literal notranslate"><span class="pre">int64</span></code>, which does not accept null values).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_curated</span><span class="p">[</span><span class="s2">&quot;PUBCHEM_SID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_curated</span><span class="p">[</span><span class="s2">&quot;PUBCHEM_SID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;Int64&quot;</span><span class="p">)</span>
<span class="n">df_curated</span><span class="p">[</span><span class="s2">&quot;PUBCHEM_CID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_curated</span><span class="p">[</span><span class="s2">&quot;PUBCHEM_CID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;Int64&quot;</span><span class="p">)</span>
<span class="n">df_curated</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>PUBCHEM_RESULT_TAG</th>
      <th>PUBCHEM_SID</th>
      <th>PUBCHEM_CID</th>
      <th>PUBCHEM_EXT_DATASOURCE_SMILES</th>
      <th>PUBCHEM_ACTIVITY_OUTCOME</th>
      <th>PUBCHEM_ACTIVITY_SCORE</th>
      <th>PUBCHEM_ACTIVITY_URL</th>
      <th>PUBCHEM_ASSAYDATA_COMMENT</th>
      <th>Activity Summary</th>
      <th>Antagonist Activity</th>
      <th>Antagonist Potency (uM)</th>
      <th>Antagonist Efficacy (%)</th>
      <th>Viability Activity</th>
      <th>Viability Potency (uM)</th>
      <th>Viability Efficacy (%)</th>
      <th>Sample Source</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>3</th>
      <td>1</td>
      <td>144203552</td>
      <td>12850184</td>
      <td>C(C(=O)[C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O.C(...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>4</th>
      <td>2</td>
      <td>144203553</td>
      <td>89753</td>
      <td>C([C@H]([C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O)O...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>5</th>
      <td>3</td>
      <td>144203554</td>
      <td>9403</td>
      <td>C[C@]12CC[C@H]3[C@H]([C@@H]1CC[C@@H]2OC(=O)CCC...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>6</th>
      <td>4</td>
      <td>144203555</td>
      <td>13218779</td>
      <td>C[C@@]12CC[C@@H](C1(C)C)C[C@H]2OC(=O)CSC#N</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>7</th>
      <td>5</td>
      <td>144203556</td>
      <td>142766</td>
      <td>C1=CC=C(C(=C1)C(=O)O)O.C1=CC2=C(C(=C1)O)N=CC=C2</td>
      <td>Inconclusive</td>
      <td>25.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inconclusive antagonist (cytotoxic)</td>
      <td>active antagonist</td>
      <td>15.5454</td>
      <td>-115.803</td>
      <td>active antagonist</td>
      <td>14.9601</td>
      <td>-76.8218</td>
      <td>NCI</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Check to see if there are any null (NaN) values.</span>
<span class="n">df_curated</span><span class="p">[[</span><span class="s2">&quot;PUBCHEM_SID&quot;</span><span class="p">,</span> <span class="s2">&quot;PUBCHEM_CID&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>PUBCHEM_SID      0
PUBCHEM_CID    149
dtype: int64
</pre></div>
</div>
</div>
</div>
<div class="alert alert-block alert-info">
<strong>Check Your Understanding</strong>
<p>Why are there 149 NaN CIDs, and no NaN SIDs?</p>
  <div style="
    background-color: #efffff;
    color: #000000;
    padding: 10px;
    border-radius: 4px;
    border: 1px solid #dddddd;
    margin-top: 10px;
  ">
<details>
    <summary>Answer</summary>
Not all SIDs could be resolved into a CID. That is, a depositor submits a substance, but PubChem's standarization process may not have been able to resolve that substance into a identified compound with known structure. Thus the value remains null.
</details>
</div>
</div></section>
<section id="save-versioned-dataframe-as-csv">
<h2>2.2 Save Versioned DataFrame as CSV<a class="headerlink" href="#save-versioned-dataframe-as-csv" title="Link to this heading">#</a></h2>
<p>You now want to save this as a file in your curated folder, but this file you will be working on and experimenting with, and so you want to use the versioning helper function. You could just write the inline command <code class="docutils literal notranslate"><span class="pre">df_raw.to_csv(&quot;Path-To-Directory/filename.csv&quot;,</span> <span class="pre">index=False)</span></code> but we are developing a project that can have many files and so need to set up protocols that use path objects and provide dates and versions within file name.</p>
<section id="helper-function-make-filename">
<h3>2.2.1 Helper Function: make_filename<a class="headerlink" href="#helper-function-make-filename" title="Link to this heading">#</a></h3>
<p>In the following code cell we are going to use the helper function <code class="docutils literal notranslate"><span class="pre">make_filename</span></code> to save this data with a dated version number appended to the file name. Like the <code class="docutils literal notranslate"><span class="pre">make_fixed_filename</span></code> we will import this from the cinf26pk package. If you do not have access to the package you can substitute it’s code in the green box below for the import statement in the code cell where it is called from.</p>
<div class="alert alert-block alert-success">
<strong>Helper Function: make_filename</strong>
<p>The following file was imported from the class package and used in the preceeding script</p>
<div style="
    background-color: #efffff;
    color: #000000;
    padding: 10px;
    border-radius: 4px;
    border: 1px solid #dddddd;
    margin-top: 10px;
  ">
<details>
    <summary>Code</summary>
 <p>The `make_filename` helper function was created in the class package pvhelpfct.py.  It is for for versioning file names with a date stamp.  This is useful when you are working up data as it allows you to save multiple versions of the file as you alter it.  That is, you may remove something, than realize you should not have removed it, and you can go back and recover the original file.</p>
<pre><code>
from pathlib import Path
from datetime import datetime
<p>def make_filename(prefix: str, ext: str = “csv”, folder: str = “downloads”) -&gt; str:
“””
Create a unique versioned filename of the form:
prefix_YYYYMMDD_vN.ext
“””
folder_path = Path(folder)
folder_path.mkdir(exist_ok=True)</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>date_str = datetime.now().strftime(&quot;%Y%m%d&quot;)
base = f&quot;{prefix}_{date_str}&quot;

existing = [
    f for f in folder_path.iterdir()
    if f.name.startswith(base) and f.suffix == f&quot;.{ext}&quot;
]

version = len(existing) + 1
filename = f&quot;{base}_v{version}.{ext}&quot;

return str(folder_path / filename)
</pre></div>
</div>
<p></code></pre></p>
</details>
</div>
</div>
<p><strong>Note: For the following code cell to generate the above directory structure your notebook must be in the 10_SupervisedML/notebooks folder.</strong> This cell will generate the curated folder within the /data/AID743139 path you had created earlier.  If you are working with a different bioassay you will need to change the AID to the new one.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">cinf26pk.core</span><span class="w"> </span><span class="kn">import</span> <span class="n">filenames</span>

<span class="c1"># generate path</span>
<span class="n">PROJECT_ROOT</span> <span class="o">=</span> <span class="n">Path</span><span class="o">.</span><span class="n">cwd</span><span class="p">()</span> <span class="c1"># project root (10_SupervisedML)</span>
<span class="n">CUR</span> <span class="o">=</span> <span class="n">PROJECT_ROOT</span> <span class="o">/</span> <span class="s2">&quot;data&quot;</span> <span class="o">/</span> <span class="s2">&quot;AID743139&quot;</span> <span class="o">/</span> <span class="s2">&quot;curated&quot;</span>
<span class="n">CUR</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Create deterministic filename with date stamp</span>
<span class="n">fname</span> <span class="o">=</span> <span class="n">filenames</span><span class="o">.</span><span class="n">make_filename</span><span class="p">(</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;AID743139_pc_curated&quot;</span><span class="p">,</span>
    <span class="n">ext</span><span class="o">=</span><span class="s2">&quot;csv&quot;</span><span class="p">,</span>
    <span class="n">folder</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">CUR</span><span class="p">)</span>
<span class="p">)</span>

<span class="c1"># Save raw dataset</span>
<span class="n">df_curated</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">CUR</span><span class="o">/</span><span class="n">fname</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[Saved] Curated PubChem data → </span><span class="si">{</span><span class="n">fname</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[Saved] Curated PubChem data → /home/rebelford/jupyterbooks/cinf26book/content/modules/10_SupervisedML/data/AID743139/curated/AID743139_pc_curated_20260118_v1.csv
</pre></div>
</div>
</div>
</div>
</section>
<section id="reload-saved-file-into-dataframe">
<h3>2.2.2 Reload saved file into dataframe<a class="headerlink" href="#reload-saved-file-into-dataframe" title="Link to this heading">#</a></h3>
<p>Load your file into a new dataframe called <strong>df_curated</strong>, and use the path variable for this new path, in contrast to the string.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Write code here: (see green box below for guidance)</span>
<span class="c1"># Load your saved csv file into a new pandas dataframe called df_curated</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="c1">#from pathlib import Path</span>

<span class="c1">#PROJECT_ROOT = Path.cwd() # project root (10_SupervisedML)</span>
<span class="c1">#CUR = PROJECT_ROOT / &quot;data&quot; / &quot;AID743139&quot; / &quot;curated&quot;</span>

<span class="n">df_curated</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">CUR</span> <span class="o">/</span> <span class="s2">&quot;AID743139_pc_curated_20260118_v1.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="alert alert-block alert-success">
<strong>Pandas Check</strong>
<p>How do you open the curated csv file in Pandas?</p>
  <div style="
    background-color: #efffff;
    color: #000000;
    padding: 10px;
    border-radius: 4px;
    border: 1px solid #dddddd;
    margin-top: 10px;
  ">
<details>
    <summary>Answer</summary>
There are two approaches, use the predefined path variable, or write the entire path as a string. For the first to work you must define the path, which was done in a preceding cell.  Of course you must use the filename of the file you saved.  If you restarted the kernel you need to redefine the path by uncommenting the commented lines of code.
<pre><code>
import pandas as pd
#from pathlib import Path
<p>#PROJECT_ROOT = Path.cwd() # project root (10_SupervisedML)
#CUR = PROJECT_ROOT / “data” / “AID743139” / “curated”</p>
<p>df_curated = pd.read_csv(CUR / “AID743139_pc_curated_20260118_v1.csv”)
</code></pre></p>
</div>
<p><strong>Inspect by displaying head</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_curated</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>PUBCHEM_RESULT_TAG</th>
      <th>PUBCHEM_SID</th>
      <th>PUBCHEM_CID</th>
      <th>PUBCHEM_EXT_DATASOURCE_SMILES</th>
      <th>PUBCHEM_ACTIVITY_OUTCOME</th>
      <th>PUBCHEM_ACTIVITY_SCORE</th>
      <th>PUBCHEM_ACTIVITY_URL</th>
      <th>PUBCHEM_ASSAYDATA_COMMENT</th>
      <th>Activity Summary</th>
      <th>Antagonist Activity</th>
      <th>Antagonist Potency (uM)</th>
      <th>Antagonist Efficacy (%)</th>
      <th>Viability Activity</th>
      <th>Viability Potency (uM)</th>
      <th>Viability Efficacy (%)</th>
      <th>Sample Source</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>1</td>
      <td>144203552</td>
      <td>12850184.0</td>
      <td>C(C(=O)[C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O.C(...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.000</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0000</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2</td>
      <td>144203553</td>
      <td>89753.0</td>
      <td>C([C@H]([C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O)O...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.000</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0000</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>2</th>
      <td>3</td>
      <td>144203554</td>
      <td>9403.0</td>
      <td>C[C@]12CC[C@H]3[C@H]([C@@H]1CC[C@@H]2OC(=O)CCC...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.000</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0000</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4</td>
      <td>144203555</td>
      <td>13218779.0</td>
      <td>C[C@@]12CC[C@@H](C1(C)C)C[C@H]2OC(=O)CSC#N</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.000</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0000</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5</td>
      <td>144203556</td>
      <td>142766.0</td>
      <td>C1=CC=C(C(=C1)C(=O)O)O.C1=CC2=C(C(=C1)O)N=CC=C2</td>
      <td>Inconclusive</td>
      <td>25.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inconclusive antagonist (cytotoxic)</td>
      <td>active antagonist</td>
      <td>15.5454</td>
      <td>-115.803</td>
      <td>active antagonist</td>
      <td>14.9601</td>
      <td>-76.8218</td>
      <td>NCI</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="alert alert-block alert-info">
<strong>Check Your Understanding</strong>
<p>Notice when you reload the dataframe from the csv the CIDs were converted back to floats, while the SIDs stayed as integers.  Can you explain what is going on?</p>
  <div style="
    background-color: #efffff;
    color: #000000;
    padding: 10px;
    border-radius: 4px;
    border: 1px solid #dddddd;
    margin-top: 10px;
  ">
<details>
    <summary>Explanation</summary>
CSV is plain-text formated and stores values only.  When Pandas opens a CSV file  it tries to identify the data type, but any column that has missing values it inserts NaN and inferres the column as a float. Since there were zero NaNs for the SIDS, those values were inferred as integers, but since there were 149 missing CIDs, those values were inferred as floats. This shows that we are too early in the pipeline to convert CIDs to integers, and we will revisit this later.
</details>
</div>
</div>
</section>
</section>
<section id="inspect-column-names">
<h2>2.3 Inspect column names<a class="headerlink" href="#inspect-column-names" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_curated</span><span class="o">.</span><span class="n">columns</span> <span class="c1">#print out the column names. </span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Index([&#39;PUBCHEM_RESULT_TAG&#39;, &#39;PUBCHEM_SID&#39;, &#39;PUBCHEM_CID&#39;,
       &#39;PUBCHEM_EXT_DATASOURCE_SMILES&#39;, &#39;PUBCHEM_ACTIVITY_OUTCOME&#39;,
       &#39;PUBCHEM_ACTIVITY_SCORE&#39;, &#39;PUBCHEM_ACTIVITY_URL&#39;,
       &#39;PUBCHEM_ASSAYDATA_COMMENT&#39;, &#39;Activity Summary&#39;, &#39;Antagonist Activity&#39;,
       &#39;Antagonist Potency (uM)&#39;, &#39;Antagonist Efficacy (%)&#39;,
       &#39;Viability Activity&#39;, &#39;Viability Potency (uM)&#39;,
       &#39;Viability Efficacy (%)&#39;, &#39;Sample Source&#39;],
      dtype=&#39;object&#39;)
</pre></div>
</div>
</div>
</div>
<section id="clean-column-names">
<h3>2.3.1 Clean column names<a class="headerlink" href="#clean-column-names" title="Link to this heading">#</a></h3>
<p>Some column names in this data frame are long and contain white spaces and special characters.  Rename the columns (no spaces or special characters except for the “_” character)  using a python dictionary as a map.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">col_names_map</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;PUBCHEM_RESULT_TAG&#39;</span> <span class="p">:</span> <span class="s1">&#39;pc_result_tag&#39;</span><span class="p">,</span> 
                 <span class="s1">&#39;PUBCHEM_SID&#39;</span> <span class="p">:</span> <span class="s1">&#39;sid&#39;</span><span class="p">,</span> 
                 <span class="s1">&#39;PUBCHEM_CID&#39;</span> <span class="p">:</span> <span class="s1">&#39;cid&#39;</span><span class="p">,</span>
                 <span class="s1">&#39;PUBCHEM_ACTIVITY_OUTCOME&#39;</span> <span class="p">:</span> <span class="s1">&#39;activity_outcome&#39;</span><span class="p">,</span> 
                 <span class="s1">&#39;PUBCHEM_ACTIVITY_SCORE&#39;</span> <span class="p">:</span> <span class="s1">&#39;activity_score&#39;</span><span class="p">,</span>
                 <span class="s1">&#39;PUBCHEM_ACTIVITY_URL&#39;</span> <span class="p">:</span> <span class="s1">&#39;activity_url&#39;</span><span class="p">,</span> 
                 <span class="s1">&#39;PUBCHEM_ASSAYDATA_COMMENT&#39;</span> <span class="p">:</span> <span class="s1">&#39;assay_data_comment&#39;</span><span class="p">,</span> 
                 <span class="s1">&#39;Activity Summary&#39;</span> <span class="p">:</span> <span class="s1">&#39;activity_summary&#39;</span><span class="p">,</span>
                 <span class="s1">&#39;Antagonist Activity&#39;</span> <span class="p">:</span> <span class="s1">&#39;antagonist_activity&#39;</span><span class="p">,</span> 
                 <span class="s1">&#39;Antagonist Potency (uM)&#39;</span> <span class="p">:</span> <span class="s1">&#39;antagonist_potency&#39;</span><span class="p">,</span> 
                 <span class="s1">&#39;Antagonist Efficacy (%)&#39;</span> <span class="p">:</span> <span class="s1">&#39;antagonist_efficacy&#39;</span><span class="p">,</span>
                 <span class="s1">&#39;Viability Activity&#39;</span> <span class="p">:</span> <span class="s1">&#39;viability_activity&#39;</span><span class="p">,</span> 
                 <span class="s1">&#39;Viability Potency (uM)&#39;</span> <span class="p">:</span> <span class="s1">&#39;viability_potency&#39;</span><span class="p">,</span>
                 <span class="s1">&#39;Viability Efficacy (%)&#39;</span> <span class="p">:</span> <span class="s1">&#39;viability_efficacy&#39;</span><span class="p">,</span> 
                 <span class="s1">&#39;Sample Source&#39;</span> <span class="p">:</span> <span class="s1">&#39;sample_source&#39;</span> <span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_curated</span> <span class="o">=</span> <span class="n">df_curated</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="n">col_names_map</span><span class="p">)</span>
<span class="n">df_curated</span><span class="o">.</span><span class="n">columns</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Index([&#39;pc_result_tag&#39;, &#39;sid&#39;, &#39;cid&#39;, &#39;PUBCHEM_EXT_DATASOURCE_SMILES&#39;,
       &#39;activity_outcome&#39;, &#39;activity_score&#39;, &#39;activity_url&#39;,
       &#39;assay_data_comment&#39;, &#39;activity_summary&#39;, &#39;antagonist_activity&#39;,
       &#39;antagonist_potency&#39;, &#39;antagonist_efficacy&#39;, &#39;viability_activity&#39;,
       &#39;viability_potency&#39;, &#39;viability_efficacy&#39;, &#39;sample_source&#39;],
      dtype=&#39;object&#39;)
</pre></div>
</div>
</div>
</div>
<div class="alert alert-block alert-info">
<strong>Check your understanding</strong>
<p>Why is it important to rename the column names without spaces or special characters?</p>
  <div style="
    background-color: #efffff;
    color: #000000;
    padding: 10px;
    border-radius: 4px;
    border: 1px solid #dddddd;
    margin-top: 10px;
  ">
<details>
    <summary>Answer</summary>
<p>Renaming columns to remove spaces and special characters creates clean, consistent labels that are easier to use in Python. We do this by creating a dictionary that maps each original column name to a standardized version and then passing it to df.rename(columns=…). This ensures that columns can be accessed reliably with either bracket notation or dot notation (e.g., df.activity_score and df.column_name) without causing syntax errors. Clean names also improve compatibility with functions, plotting libraries, and automated workflows that expect simple identifiers. In short, renaming columns makes your code more readable, robust, and less error-prone.</p>
</details>
</div>
</div>
<div class="alert alert-block alert-info"> 
<strong>Data Dive:</strong> 
    <p>The following pop-up will describe each column in the dataframe. It should be noted that 
<a href="https://pubchem.ncbi.nlm.nih.gov/bioassay/743139" target="_blank" rel="noopener noreferrer">Bioassay 743139</a> is a summary assayfor measuring inhibition of aramatase, and so a positive assay result means a chemical is an inhibitor (antagonist) of the enzyme.  The assay also measures agonist activity, but these are not considered active in this assay, but as inconlusive. In other words, only antagonists are considered active in this assay.
    </p>
  <div style="
    background-color: #efffff;
    color: #000000;
    padding: 10px;
    border-radius: 4px;
    border: 1px solid #dddddd;
    margin-top: 10px;
  ">
<details>
    <summary>Explanation of the data</summary>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Column Name</p></th>
<th class="head"><p>Meaning / Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>pc_result_tag</strong></p></td>
<td><p>Internal PubChem result index. A simple numeric tag identifying the order of assay records. Not biologically meaningful.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>sid</strong> (Substance ID)</p></td>
<td><p>PubChem <strong>Substance ID</strong>: identifies the submitted sample tested in the assay (may map to multiple CIDs).</p></td>
</tr>
<tr class="row-even"><td><p><strong>cid</strong> (Compound ID)</p></td>
<td><p>PubChem <strong>Compound ID</strong>: normalized chemical structure identifier used across PubChem. Common structure-level identifier for cheminformatics.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>PUBCHEM_EXT_DATASOURCE_SMILES</strong></p></td>
<td><p>A SMILES string representing the chemical structure as submitted by the data source. May differ from canonical PubChem SMILES.</p></td>
</tr>
<tr class="row-even"><td><p><strong>activity_outcome</strong></p></td>
<td><p><strong>High-level BioAssay classification</strong>: <em>Active</em>, <em>Inactive</em>, or <em>Inconclusive</em>. This is the main result label for each tested substance.   This is a summary table and this result depends on a biological threshold of the actual assays.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>activity_score</strong></p></td>
<td><p>Numeric score assigned by PubChem summarizing activity strength; depends on assay scoring rules. Often 0 = inactive, higher = stronger activity.</p></td>
</tr>
<tr class="row-even"><td><p><strong>activity_url</strong></p></td>
<td><p>URL linking to additional PubChem assay information for this result. Frequently missing (NaN) in simplified datasets.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>assay_data_comment</strong></p></td>
<td><p>Optional textual note describing assay-specific observations (rare in many datasets).</p></td>
</tr>
<tr class="row-even"><td><p><strong>activity_summary</strong></p></td>
<td><p><strong>More detailed subtype of activity_outcome</strong> (e.g., <em>active antagonist</em>, <em>inconclusive agonist</em>, <em>inconclusive antagonist (cytotoxic)</em>). Provides mechanism- or phenotype-level interpretation.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>antagonist_activity</strong></p></td>
<td><p>Qualitative label identifying whether the compound acted as an <strong>antagonist</strong> in the assay (active/inactive).</p></td>
</tr>
<tr class="row-even"><td><p><strong>antagonist_potency</strong></p></td>
<td><p>Numeric potency value for antagonist activity (often in μM). Indicates concentration required for effect.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>antagonist_efficacy</strong></p></td>
<td><p>Percent efficacy of antagonist behavior. Measures the magnitude of response relative to assay controls.</p></td>
</tr>
<tr class="row-even"><td><p><strong>viability_activity</strong></p></td>
<td><p>Qualitative label describing <strong>cell viability response</strong> (active/inactive). Helps detect cytotoxic compounds.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>viability_potency</strong></p></td>
<td><p>Potency value for viability-related effects (e.g., concentration at which viability decreases).</p></td>
</tr>
<tr class="row-even"><td><p><strong>viability_efficacy</strong></p></td>
<td><p>Percent efficacy for viability effects (extent of cell killing or protection).</p></td>
</tr>
<tr class="row-odd"><td><p><strong>sample_source</strong></p></td>
<td><p>Origin of the tested sample (e.g., NCI). Refers to the organization or project that submitted the substance to PubChem.</p></td>
</tr>
</tbody>
</table>
</div>
</details>
</div>
</div>
</section>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="inspect-activity-outcomes">
<h1>3. Inspect Activity Outcomes<a class="headerlink" href="#inspect-activity-outcomes" title="Link to this heading">#</a></h1>
<p>First, we need to understand the structure of our data, particularly the activity class of the tested compounds, since our goal is to develop a model that classifies small molecules based on their activity against a target. This information is found in the <strong>activity_outcome</strong> and <strong>activity_summary</strong> columns of the dataframe. The activity outcomes have three values, Active, Inactive and Inconclusive.</p>
<p>To explore this, we use the <code class="docutils literal notranslate"><span class="pre">groupby</span></code> method, which splits the data into groups based on one or more column values, enabling us to perform calculations on each group independently. In the code cell below, we group by <strong>activity_outcome</strong> and use the <code class="docutils literal notranslate"><span class="pre">count</span></code> method to determine how many entries fall into each outcome category:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_curated</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;activity_outcome&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>pc_result_tag</th>
      <th>sid</th>
      <th>cid</th>
      <th>PUBCHEM_EXT_DATASOURCE_SMILES</th>
      <th>activity_score</th>
      <th>activity_url</th>
      <th>assay_data_comment</th>
      <th>activity_summary</th>
      <th>antagonist_activity</th>
      <th>antagonist_potency</th>
      <th>antagonist_efficacy</th>
      <th>viability_activity</th>
      <th>viability_potency</th>
      <th>viability_efficacy</th>
      <th>sample_source</th>
    </tr>
    <tr>
      <th>activity_outcome</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Active</th>
      <td>379</td>
      <td>379</td>
      <td>378</td>
      <td>378</td>
      <td>379</td>
      <td>0</td>
      <td>0</td>
      <td>379</td>
      <td>379</td>
      <td>378</td>
      <td>379</td>
      <td>379</td>
      <td>115</td>
      <td>359</td>
      <td>379</td>
    </tr>
    <tr>
      <th>Inactive</th>
      <td>7562</td>
      <td>7562</td>
      <td>7466</td>
      <td>7466</td>
      <td>7562</td>
      <td>0</td>
      <td>0</td>
      <td>7562</td>
      <td>7562</td>
      <td>0</td>
      <td>7562</td>
      <td>7562</td>
      <td>324</td>
      <td>7449</td>
      <td>7562</td>
    </tr>
    <tr>
      <th>Inconclusive</th>
      <td>2545</td>
      <td>2545</td>
      <td>2493</td>
      <td>2493</td>
      <td>2545</td>
      <td>0</td>
      <td>0</td>
      <td>2545</td>
      <td>2545</td>
      <td>2111</td>
      <td>2136</td>
      <td>2545</td>
      <td>1206</td>
      <td>2450</td>
      <td>2545</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="alert alert-block alert-info">
<strong>Check your understanding</strong>
<p>1. Explain this line of code: `df_raw.groupby(['activity_outcome']).count()`</p>
  <div style="
    background-color: #efffff;
    color: #000000;
    padding: 10px;
    border-radius: 4px;
    border: 1px solid #dddddd;
    margin-top: 10px;
  ">
<details>
    <summary>Answer</summary>
<ol class="arabic simple">
<li><p>This represents two chained Pandas methods;  .groupby() followed by .count().</p>
<ul class="simple">
<li><p><strong>.groupby(‘activity_outcome’)</strong> splits the DataFrame into groups based on the category labels in the activity_outcome column; namely “Active,” “Inactive,” and “Inconclusive.”</p></li>
<li><p><strong>.count()</strong> then counts how many non-missing values appear in each of the other columns within each category. This provides a quick summary of how many compounds fall into each activity class and reveals whether key fields (such as CID or SMILES) contain missing data within any category.</p></li>
</ul>
</li>
</ol>
</details>
</div>
<p>2. How many actives, inactives and inconclusives compounds are there in the dataset based on the <b>activity_outcome</b> column?</p>
  <div style="
    background-color: #efffff;
    color: #000000;
    padding: 10px;
    border-radius: 4px;
    border: 1px solid #dddddd;
    margin-top: 10px;
  ">
<details>
    <summary>Answer</summary>
2. Based on the data in the activity_outcome column, there are 379 actives, 7562 inactives, and 2545 inconclusive compounds in the dataset.
</details>
</div>
<p>3. Scroll through the rest of the data. Compare these numbers to other columns such as CID and SMILES. What problems might arise if these counts are not consistent?</p>
  <div style="
    background-color: #efffff;
    color: #000000;
    padding: 10px;
    border-radius: 4px;
    border: 1px solid #dddddd;
    margin-top: 10px;
  ">
<details>
    <summary>Answer</summary>
3. If the counts of compounds differ across columns (for example, if a compound has an activity_outcome but is missing a CID or SMILES), it means important information is missing. This causes problems such as:  
<ul>
  <li>Compounds with missing CIDs/SMILES cannot be used for structural analysis, descriptor generation, or machine learning.</li>
  <li>The dataset can become biased if missing data are unevenly distributed across activity classes, which can affect model performance and interpretation.</li>
  <li>Results become harder to interpret because activity labels may exist without the structural or potency data needed to contextualize them.</li>
</ul>
</details>
</div>
</div>
<section id="inspect-activity-summaries">
<h2>3.1 Inspect Activity summaries<a class="headerlink" href="#inspect-activity-summaries" title="Link to this heading">#</a></h2>
<p>If we group by both <strong>activity_outcome</strong> and <strong>activity_summary</strong> we get more insight about the data. Particularly, it reveals information that will indicate that there are subcategories for “<strong>Inconclusive</strong>”, giving us a better understanding of how thse results are classified.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_curated</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;activity_outcome&#39;</span><span class="p">,</span><span class="s1">&#39;activity_summary&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th>pc_result_tag</th>
      <th>sid</th>
      <th>cid</th>
      <th>PUBCHEM_EXT_DATASOURCE_SMILES</th>
      <th>activity_score</th>
      <th>activity_url</th>
      <th>assay_data_comment</th>
      <th>antagonist_activity</th>
      <th>antagonist_potency</th>
      <th>antagonist_efficacy</th>
      <th>viability_activity</th>
      <th>viability_potency</th>
      <th>viability_efficacy</th>
      <th>sample_source</th>
    </tr>
    <tr>
      <th>activity_outcome</th>
      <th>activity_summary</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Active</th>
      <th>active antagonist</th>
      <td>379</td>
      <td>379</td>
      <td>378</td>
      <td>378</td>
      <td>379</td>
      <td>0</td>
      <td>0</td>
      <td>379</td>
      <td>378</td>
      <td>379</td>
      <td>379</td>
      <td>115</td>
      <td>359</td>
      <td>379</td>
    </tr>
    <tr>
      <th>Inactive</th>
      <th>inactive</th>
      <td>7562</td>
      <td>7562</td>
      <td>7466</td>
      <td>7466</td>
      <td>7562</td>
      <td>0</td>
      <td>0</td>
      <td>7562</td>
      <td>0</td>
      <td>7562</td>
      <td>7562</td>
      <td>324</td>
      <td>7449</td>
      <td>7562</td>
    </tr>
    <tr>
      <th rowspan="6" valign="top">Inconclusive</th>
      <th>active agonist</th>
      <td>612</td>
      <td>612</td>
      <td>571</td>
      <td>571</td>
      <td>612</td>
      <td>0</td>
      <td>0</td>
      <td>612</td>
      <td>612</td>
      <td>612</td>
      <td>612</td>
      <td>60</td>
      <td>590</td>
      <td>612</td>
    </tr>
    <tr>
      <th>inconclusive</th>
      <td>44</td>
      <td>44</td>
      <td>44</td>
      <td>44</td>
      <td>44</td>
      <td>0</td>
      <td>0</td>
      <td>44</td>
      <td>0</td>
      <td>0</td>
      <td>44</td>
      <td>19</td>
      <td>42</td>
      <td>44</td>
    </tr>
    <tr>
      <th>inconclusive agonist</th>
      <td>414</td>
      <td>414</td>
      <td>409</td>
      <td>409</td>
      <td>414</td>
      <td>0</td>
      <td>0</td>
      <td>414</td>
      <td>212</td>
      <td>223</td>
      <td>414</td>
      <td>12</td>
      <td>397</td>
      <td>414</td>
    </tr>
    <tr>
      <th>inconclusive agonist (cytotoxic)</th>
      <td>59</td>
      <td>59</td>
      <td>59</td>
      <td>59</td>
      <td>59</td>
      <td>0</td>
      <td>0</td>
      <td>59</td>
      <td>41</td>
      <td>45</td>
      <td>59</td>
      <td>59</td>
      <td>59</td>
      <td>59</td>
    </tr>
    <tr>
      <th>inconclusive antagonist</th>
      <td>367</td>
      <td>367</td>
      <td>364</td>
      <td>364</td>
      <td>367</td>
      <td>0</td>
      <td>0</td>
      <td>367</td>
      <td>227</td>
      <td>230</td>
      <td>367</td>
      <td>8</td>
      <td>313</td>
      <td>367</td>
    </tr>
    <tr>
      <th>inconclusive antagonist (cytotoxic)</th>
      <td>1049</td>
      <td>1049</td>
      <td>1046</td>
      <td>1046</td>
      <td>1049</td>
      <td>0</td>
      <td>0</td>
      <td>1049</td>
      <td>1019</td>
      <td>1026</td>
      <td>1049</td>
      <td>1048</td>
      <td>1049</td>
      <td>1049</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Note, sometimes the above display sort of creates a UI illusion.  Run the next code and you will see that activity_outcome Active and Inactive have only one type of activity_summary, and Inconclusive has many, of which one is active agonist.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">outcome</span><span class="p">,</span> <span class="n">summaries</span> <span class="ow">in</span> <span class="p">(</span>
    <span class="n">df_curated</span>
    <span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;activity_outcome&#39;</span><span class="p">)[</span><span class="s1">&#39;activity_summary&#39;</span><span class="p">]</span>
    <span class="o">.</span><span class="n">unique</span><span class="p">()</span>
    <span class="o">.</span><span class="n">items</span><span class="p">()</span>
<span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="si">{</span><span class="n">outcome</span><span class="si">}</span><span class="s2">:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">summaries</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  - </span><span class="si">{</span><span class="n">s</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Active:
  - active antagonist

Inactive:
  - inactive

Inconclusive:
  - active agonist
  - inconclusive
  - inconclusive agonist
  - inconclusive agonist (cytotoxic)
  - inconclusive antagonist
  - inconclusive antagonist (cytotoxic)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;there are&#39;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">df_curated</span><span class="p">),</span><span class="s1">&#39;compounds in the dataframe that are defined as active antagonist, inactive, or inconclusive.&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>there are 10486 compounds in the dataframe that are defined as active antagonist, inactive, or inconclusive.
</pre></div>
</div>
</div>
</div>
<p>Now, we can see that, in the <strong>activity_summary</strong> column, the inconclusive compounds are further classified into subclasses, which include:</p>
<ul class="simple">
<li><p><strong>active agonist</strong></p></li>
<li><p>inconclusive</p></li>
<li><p>inconclusive agonist</p></li>
<li><p>inconclusive antagonist</p></li>
<li><p>inconclusive agonist (cytotoxic)</p></li>
<li><p>inconclusive antagonist (cytotoxic)</p></li>
</ul>
<p>As indicated by the title of this <a class="reference external" href="https://pubchem.ncbi.nlm.nih.gov/bioassay/743139">PubChem assay record (AID 743139)</a>, the purpose of the experiment was to identify aromatase inhibitors. For this reason, the original submitter labeled all compounds showing antagonist behavior as active in the activity_outcome column. The assay also detected 612 agonists, but these were marked as inconclusive because the submitter defined activity strictly in terms of inhibitory, not stimulatory, effects. In this dataset, inactive compounds are those that show neither agonist nor antagonist behavior.</p>
<p>It is important to remember that “active” and “inactive” are not universal labels; their meaning depends entirely on the assay design and the submitter’s interpretation. For our purposes, we redefine these categories to support a binary classification task: any compound that changes the function of the target, whether by increasing or decreasing its activity, is considered active, and compounds that do not alter the target are considered inactive. In short, we are grouping all function-altering interactions together as active and treating the remainder as inactive, so the model can focus on distinguishing compounds that affect the enzyme from those that do not.</p>
<div class="alert alert-block alert-info">
<p><strong>Check Your Understanding: Defining “Active” as a Modeling Decision</strong></p>
<p>
As we prepare to collapse assay outcomes into a binary classification for model building (active vs inactive), we must decide how to handle compounds labeled as inconclusive. In this dataset, there are 379 active antagonists, 7,562 inactive compounds, and 2,545 inconclusive compounds. Notably, 612 of the inconclusive compounds are labeled as active agonists, substantially more than the number of active antagonists. Rather than asking whether a particular choice is correct, this discussion is intended to surface the consequences of the choices we make and to emphasize the importance of recognizing that a choice is being made at all.
</p>
  <div style="
    background-color: #efffff;
    color: #000000;
    padding: 10px;
    border-radius: 4px;
    border: 1px solid #dddddd;
    margin-top: 10px;
  ">
<details>
    <summary>Discussion Points</summary>
<p><strong>The deeper lesson: recognizing that a decision was made</strong></p>
<ul>
  <li>At what point in the workflow does the definition of “active” enter the modeling process?</li>
  <li>How does collapsing rich experimental outcomes into a binary label necessarily simplify the biology?</li>
  <li>What assumptions are we making about mechanism, relevance, and similarity when we redefine labels?</li>
  <li>How might a different research goal lead to a different but equally reasonable decision using the same data?</li>
</ul>
<p><strong>Why documentation may matter more than the decision itself</strong></p>
<ul>
  <li>Could someone reuse this model without realizing how “active” was defined?</li>
  <li>How would performance metrics be interpreted differently if agonists were excluded rather than included?</li>
  <li>How does documenting this choice enable reproducibility, comparison, and critical evaluation?</li>
  <li>What risks arise when modeling assumptions are implicit rather than explicit?</li>
</ul>
<p><strong>Factors that influence this decision</strong></p>
<ul>
  <li>Biological considerations: agonists are clearly bioactive and demonstrate target engagement, even if the mechanism differs.</li>
  <li>Chemical considerations: agonists and antagonists may share structural features relevant to binding and recognition.</li>
  <li>Data considerations: there are more active agonists (612) than active antagonists (379), and excluding them would further reduce an already small positive class.</li>
  <li>Modeling considerations: extreme class imbalance can negatively affect learning, evaluation, and generalization.</li>
</ul>
<p><strong>Consequences of including active agonists as “active”</strong></p>
<ul>
  <li>The model becomes a general bioactivity classifier rather than a mechanism-specific antagonist detector.</li>
  <li>Predictions answer the question “is this compound active against the target?” rather than “is this compound an antagonist?”</li>
  <li>Mechanistic distinctions are deferred to later analysis rather than embedded in the initial model.</li>
</ul>
<p><strong>Consequences of excluding active agonists</strong></p>
<ul>
  <li>The active class becomes very small relative to the inactive class, increasing imbalance.</li>
  <li>Biologically active compounds are treated as negatives, potentially obscuring meaningful structure–activity relationships.</li>
  <li>The model becomes more specialized but less general and less transferable.</li>
</ul>
</details>
</div>
</div>
<p><strong>Connecting this decision to the next step</strong></p>
<ul>
  <li>In the next section, we will explicitly define active and inactive classes for binary classification.</li>
  <li>All inconclusive compounds will be removed except those labeled as active agonists.</li>
  <li>This decision should be understood as a deliberate modeling choice shaped by biological goals, data constraints, and downstream use.</li>
</ul>
<p>
The central takeaway is not that one labeling strategy is universally correct, but that supervised learning models are built on decisions that shape what they can learn, what they ignore, and how their predictions should be interpreted.
Recognizing, justifying, and documenting those decisions is a core part of responsible model building.
</p></section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="select-active-inactive-compounds-for-model-building">
<h1>4. Select active/inactive compounds for model building<a class="headerlink" href="#select-active-inactive-compounds-for-model-building" title="Link to this heading">#</a></h1>
<section id="select-active-and-inactive-compounds">
<h2>4.1 Select active and inactive compounds<a class="headerlink" href="#select-active-and-inactive-compounds" title="Link to this heading">#</a></h2>
<p>Now we want to select only the active and inactive compounds from the data frame (that is, active agonists, active antagonists, and inactives based on the “activity summary” column). This code filters the DataFrame so that only rows labeled as active agonist, active antagonist, or inactive are kept, using the | operator to combine multiple Boolean conditions with “OR.” This removes all inconclusive entries, ensuring that the resulting DataFrame contains only compounds with clear activity classifications suitable for model training.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a new dataframe with active agonists, active antagnonists and inactives</span>
<span class="c1"># remove any molecules that are inconclusive and not active agonist</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">df_curated</span><span class="p">[</span> <span class="p">(</span><span class="n">df_curated</span><span class="p">[</span><span class="s1">&#39;activity_summary&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;active agonist&#39;</span> <span class="p">)</span> <span class="o">|</span> 
             <span class="p">(</span><span class="n">df_curated</span><span class="p">[</span><span class="s1">&#39;activity_summary&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;active antagonist&#39;</span> <span class="p">)</span> <span class="o">|</span>
             <span class="p">(</span><span class="n">df_curated</span><span class="p">[</span><span class="s1">&#39;activity_summary&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;inactive&#39;</span> <span class="p">)</span> <span class="p">]</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The number of total molecules that are active or inactive is:&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>The number of total molecules that are active or inactive is: 8553
</pre></div>
</div>
</div>
</div>
<p>Since we will be obtaining structural data from PubChem, it’s important to examine how many unique PubChem Substance IDs (SIDs) and Compound IDs (CIDs) are present in our dataset. Recall the distinction:</p>
<ul class="simple">
<li><p><strong>Substances (SIDs)</strong> represent depositor-submitted records. Multiple SIDs may refer to the same chemical structure if submitted by different sources.</p></li>
<li><p><strong>Compounds (CIDs)</strong> are standardized, unique chemical structures derived by PubChem from submitted substances through a process of structure normalization and deduplication.</p></li>
</ul>
<p>Understanding how many unique SIDs and CIDs we have will help us assess redundancy and ensure we’re working with non-duplicated chemical structures when building models or visualizing data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># identify Compounds IDs and Substance IDs. </span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;total number of unique Substance IDs=&#39;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;sid&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;total number of unique Compound IDs= &#39;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;cid&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>total number of unique Substance IDs= 8553
total number of unique Compound IDs=  6858
</pre></div>
</div>
</div>
</div>
<p>This data indicates several important points:</p>
<ol class="arabic simple">
<li><p>The number of unique Substance IDs matches the number of active and inactive molecules in the dataset.</p></li>
<li><p>Not all substances have an associated Compound ID (CID), remember we identified 149 of them earlier, meaning some structures have not been standardized by PubChem.</p></li>
<li><p>Some substances may have the same CID, indicating that different submissions or assay runs may correspond to the same chemical structure.</p></li>
</ol>
<p>Since our goal is to build a model that uses chemical structure to predict biological activity, we must remove substances without CIDs as these lack standardized structural information necessary for modeling. We must also remove duplicate SIDs for the same CID, as they are the same structure. By cleaning our data based on CID and activity type, we ensure the model learns from reliable, unambiguous data where each chemical structure is uniquely associated with a single activity classification.</p>
</section>
<section id="drop-substances-without-associated-cids">
<h2>4.2 Drop substances without associated CIDs.<a class="headerlink" href="#drop-substances-without-associated-cids" title="Link to this heading">#</a></h2>
<p>First, check if there are subtances without associated CIDs. We will used the <code class="docutils literal notranslate"><span class="pre">.isna()</span></code> method to identify all rows that are missing values (not available). Note the following code cell uses <em>chaining</em> to sequentially call multiple methods on a pandas object (we chain <code class="docutils literal notranslate"><span class="pre">.isna</span></code> with <code class="docutils literal notranslate"><span class="pre">.sum()</span></code> to count how many are missing).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>  
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>pc_result_tag                       0
sid                                 0
cid                               138
PUBCHEM_EXT_DATASOURCE_SMILES     138
activity_outcome                    0
activity_score                      0
activity_url                     8553
assay_data_comment               8553
activity_summary                    0
antagonist_activity                 0
antagonist_potency               7563
antagonist_efficacy                 0
viability_activity                  0
viability_potency                8054
viability_efficacy                155
sample_source                       0
dtype: int64
</pre></div>
</div>
</div>
</div>
<div class="alert alert-block alert-info">
<strong>Check your understanding</strong>
<p>How many records lack an associated CID? Why does this number differ from the 149 NaNs identified in section 4.1? </p>
  <div style="
    background-color: #efffff;
    color: #000000;
    padding: 10px;
    border-radius: 4px;
    border: 1px solid #dddddd;
    margin-top: 10px;
  ">
<details>
    <summary>Answer</summary>
<p>138 compounds do not have a CID. There was originally 149 and thereare fewer now because we have removed all inconslusive compounds that were not active agonist.</p>
</details>
</div>
</div><p>The records without CID are not going to provide us with standardized structural information, so we will remove them from the dataframe.</p>
<p>This is accomplished by using the <code class="docutils literal notranslate"><span class="pre">.dropna()</span></code> method. Since we only want to remove rows where the CID is missing, we can use the <code class="docutils literal notranslate"><span class="pre">subset</span></code> parameter to specify we are targeting only those NA values in the <strong>‘cid’</strong> column.</p>
<p>Look at the output of the previous cell block. If we did the <code class="docutils literal notranslate"><span class="pre">dropna()</span></code> method without indicating which subset we are dropping, we could lose all of our data because the <strong>activity_url</strong> and **assay_data_comment” are null for every row in the dataframe!</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">))</span>   <span class="c1">#in previous we saw 138 rows without CID</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span> <span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;cid&#39;</span><span class="p">]</span> <span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">))</span>   <span class="c1">#this values should be 138 fewer</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>8553
8415
</pre></div>
</div>
</div>
</div>
<p>Check if the NULL values disappeared in the <strong>cid</strong> column. The value should be 0 for that column now.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>pc_result_tag                       0
sid                                 0
cid                                 0
PUBCHEM_EXT_DATASOURCE_SMILES       0
activity_outcome                    0
activity_score                      0
activity_url                     8415
assay_data_comment               8415
activity_summary                    0
antagonist_activity                 0
antagonist_potency               7467
antagonist_efficacy                 0
viability_activity                  0
viability_potency                7919
viability_efficacy                154
sample_source                       0
dtype: int64
</pre></div>
</div>
</div>
</div>
<p>Now that we have removed all CIDs that are NaN we can safely convert these to integers and not worry about the value being converted back to a float when we reload a saved csv file.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Write code here to convert CIDs to integers and print the head</span>
<span class="n">df</span><span class="p">[</span><span class="s2">&quot;cid&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;cid&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;Int64&quot;</span><span class="p">)</span>
<span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>pc_result_tag</th>
      <th>sid</th>
      <th>cid</th>
      <th>PUBCHEM_EXT_DATASOURCE_SMILES</th>
      <th>activity_outcome</th>
      <th>activity_score</th>
      <th>activity_url</th>
      <th>assay_data_comment</th>
      <th>activity_summary</th>
      <th>antagonist_activity</th>
      <th>antagonist_potency</th>
      <th>antagonist_efficacy</th>
      <th>viability_activity</th>
      <th>viability_potency</th>
      <th>viability_efficacy</th>
      <th>sample_source</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>1</td>
      <td>144203552</td>
      <td>12850184</td>
      <td>C(C(=O)[C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O.C(...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2</td>
      <td>144203553</td>
      <td>89753</td>
      <td>C([C@H]([C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O)O...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>2</th>
      <td>3</td>
      <td>144203554</td>
      <td>9403</td>
      <td>C[C@]12CC[C@H]3[C@H]([C@@H]1CC[C@@H]2OC(=O)CCC...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4</td>
      <td>144203555</td>
      <td>13218779</td>
      <td>C[C@@]12CC[C@@H](C1(C)C)C[C@H]2OC(=O)CSC#N</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>5</th>
      <td>6</td>
      <td>144203557</td>
      <td>16043</td>
      <td>CC(C)(C)C1=C(C=CC(=C1)O)O</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>NCI</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># identify Compounds IDs and Substance IDs. </span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;total number of Substance IDs=&#39;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;sid&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;total number of Compound IDs= &#39;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;cid&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>total number of Substance IDs= 8415
total number of Compound IDs=  6857
</pre></div>
</div>
</div>
</div>
<div class="alert alert-block alert-info">
<strong>Check your understanding</strong>
<p>Why can the numer of SIDs be greater than the number of CIDs? </p>
  <div style="
    background-color: #efffff;
    color: #000000;
    padding: 10px;
    border-radius: 4px;
    border: 1px solid #dddddd;
    margin-top: 10px;
  ">
<details>
    <summary>Answer</summary>
<p>The SID is related to the submitter and multiple assays could have been performed on the same chemical. This means we not only have multiple entries for the same chemical compound (CID), but it could result in conflicting activities.</p>
</details>
</div>
</div></section>
<section id="remove-cids-with-conflicting-activities">
<h2>4.3 Remove CIDs with conflicting activities<a class="headerlink" href="#remove-cids-with-conflicting-activities" title="Link to this heading">#</a></h2>
<p>This code identifies unique CIDs and checks whether multiple values exist in the <strong>activity_summary column</strong>. If conflicting activity summaries are found, the code stores the CID in the <code class="docutils literal notranslate"><span class="pre">cid_conflict</span></code> list and and records the corresponding row indices in <code class="docutils literal notranslate"><span class="pre">idx_conflict</span></code>. Finally, the total number of CIDs with conflicts and total number of rows are output.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cid_conflict</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1"># list to store CIDs with conflicting activities</span>
<span class="n">idx_conflict</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1"># list to store indices of rows with conflicting activities</span>

<span class="k">for</span> <span class="n">mycid</span> <span class="ow">in</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;cid&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span> <span class="p">:</span> <span class="c1"># iterate over each unique Compound ID</span>
    
    <span class="n">outcomes</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span> <span class="n">df</span><span class="o">.</span><span class="n">cid</span> <span class="o">==</span> <span class="n">mycid</span> <span class="p">]</span><span class="o">.</span><span class="n">activity_summary</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span> <span class="c1">#</span>
    
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">outcomes</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="p">:</span> <span class="c1"># if there are multiple unique activity summaries for this CID</span>
        
        <span class="n">idx_tmp</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">[</span> <span class="n">df</span><span class="o">.</span><span class="n">cid</span> <span class="o">==</span> <span class="n">mycid</span> <span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span> <span class="c1"># get the indices of these rows</span>
        <span class="n">idx_conflict</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">idx_tmp</span><span class="p">)</span> <span class="c1"># add these indices to the conflict list</span>
        <span class="n">cid_conflict</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mycid</span><span class="p">)</span> <span class="c1"># # add the CID to the conflict list</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">cid_conflict</span><span class="p">),</span> <span class="s2">&quot;CIDs with conflicting activities [associated with&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">idx_conflict</span><span class="p">),</span> <span class="s2">&quot;rows (SIDs).]&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">idx_conflict</span><span class="p">[:</span><span class="mi">10</span><span class="p">])</span> <span class="c1"># print first 10 CID index numbers of conflicting activities</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span># 66 CIDs with conflicting activities [associated with 150 rows (SIDs).]
[5, 5953, 6847, 49, 6127, 63, 9115, 103, 5917, 6961]
</pre></div>
</div>
</div>
</div>
<p>To examine which CIDs have conflicting activity data, we can display a portion of the dataframe using the <code class="docutils literal notranslate"><span class="pre">.loc[]</span></code> method. This allows us to select specific rows. In this case, we used the ones listed in <code class="docutils literal notranslate"><span class="pre">idx_conflict</span></code>, which correspond to compounds with inconsistent activity summaries. We’ll chain with <code class="docutils literal notranslate"><span class="pre">.head(10)</span></code> to view just the first 10 of these rows for quick inspection.</p>
<p>This helps us verify the nature of the conflicts and better understand why these entries need to be removed before modeling.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">idx_conflict</span><span class="p">,:]</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>pc_result_tag</th>
      <th>sid</th>
      <th>cid</th>
      <th>PUBCHEM_EXT_DATASOURCE_SMILES</th>
      <th>activity_outcome</th>
      <th>activity_score</th>
      <th>activity_url</th>
      <th>assay_data_comment</th>
      <th>activity_summary</th>
      <th>antagonist_activity</th>
      <th>antagonist_potency</th>
      <th>antagonist_efficacy</th>
      <th>viability_activity</th>
      <th>viability_potency</th>
      <th>viability_efficacy</th>
      <th>sample_source</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>5</th>
      <td>6</td>
      <td>144203557</td>
      <td>16043</td>
      <td>CC(C)(C)C1=C(C=CC(=C1)O)O</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0000</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0000</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>5953</th>
      <td>5954</td>
      <td>144209507</td>
      <td>16043</td>
      <td>CC(C)(C)C1=C(C=CC(=C1)O)O</td>
      <td>Active</td>
      <td>43.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>active antagonist</td>
      <td>active antagonist</td>
      <td>54.48270</td>
      <td>-73.4024</td>
      <td>inconclusive antagonist</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>SigmaAldrich</td>
    </tr>
    <tr>
      <th>6847</th>
      <td>6848</td>
      <td>144210401</td>
      <td>16043</td>
      <td>CC(C)(C)C1=C(C=CC(=C1)O)O</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0000</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0000</td>
      <td>SIGMA</td>
    </tr>
    <tr>
      <th>49</th>
      <td>50</td>
      <td>144203601</td>
      <td>443939</td>
      <td>C[C@H]1[C@H]([C@H](C[C@@H](O1)O[C@H]2C[C@@](CC...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0000</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0000</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>6127</th>
      <td>6128</td>
      <td>144209681</td>
      <td>443939</td>
      <td>C[C@H]1[C@H]([C@H](C[C@@H](O1)O[C@H]2C[C@@](CC...</td>
      <td>Active</td>
      <td>61.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>active antagonist</td>
      <td>active antagonist</td>
      <td>1.65519</td>
      <td>-115.9320</td>
      <td>active antagonist</td>
      <td>12.1763</td>
      <td>-120.5980</td>
      <td>Toronto Research</td>
    </tr>
    <tr>
      <th>63</th>
      <td>64</td>
      <td>144203615</td>
      <td>2170</td>
      <td>C1CN(CCN1)C2=NC3=CC=CC=C3OC4=C2C=C(C=C4)Cl</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0000</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0000</td>
      <td>BIOMOL</td>
    </tr>
    <tr>
      <th>9115</th>
      <td>9116</td>
      <td>144212669</td>
      <td>2170</td>
      <td>C1CN(CCN1)C2=NC3=CC=CC=C3OC4=C2C=C(C=C4)Cl</td>
      <td>Active</td>
      <td>50.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>active antagonist</td>
      <td>active antagonist</td>
      <td>16.58030</td>
      <td>-115.2020</td>
      <td>inconclusive antagonist</td>
      <td>61.1306</td>
      <td>-80.7706</td>
      <td>SIGMA</td>
    </tr>
    <tr>
      <th>103</th>
      <td>104</td>
      <td>144203655</td>
      <td>2554</td>
      <td>C1=CC=C2C(=C1)C=CC3=CC=CC=C3N2C(=O)N</td>
      <td>Inconclusive</td>
      <td>20.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>active agonist</td>
      <td>active agonist</td>
      <td>2.87255</td>
      <td>73.7025</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0000</td>
      <td>SigmaAldrich</td>
    </tr>
    <tr>
      <th>5917</th>
      <td>5918</td>
      <td>144209471</td>
      <td>2554</td>
      <td>C1=CC=C2C(=C1)C=CC3=CC=CC=C3N2C(=O)N</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0000</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0000</td>
      <td>SIGMA</td>
    </tr>
    <tr>
      <th>6961</th>
      <td>6962</td>
      <td>144210515</td>
      <td>2554</td>
      <td>C1=CC=C2C(=C1)C=CC3=CC=CC=C3N2C(=O)N</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0000</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0000</td>
      <td>SIGMA</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>We will be creating a binary classifier system distinguishing between active and inactive compounds, and will lump active antagonist with active agonist as active.  The active agonist are being pulled from the activity_summary of <em>inactive compounds</em> in the activity_outcome data, and so are being read as inactive when comparing the activity_outcome. This means we are dropping them as conflicting outcomes even though we will consider active agonist as active.</p>
<p>This should be rare as one assay is saying the compound is active antagonist (activity_outcome [Active] = active antagonist) and the other as active agonist (activity_outcome [Inconclusive]activity_summary= active agonist). The following code cell identifies these compounds and returns a python set of them with the variable name allowed_cids.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Identify conflicting CIDs using activity_summary (already computed earlier)</span>
<span class="c1"># idx_conflict comes from the earlier conflict-detection cell</span>

<span class="c1"># Define biologically allowed exceptions:</span>
<span class="c1"># CIDs that have both active antagonist and active agonist summaries</span>
<span class="n">allowed_cids</span> <span class="o">=</span> <span class="p">[]</span>

<span class="k">for</span> <span class="n">cid</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;cid&#39;</span><span class="p">):</span>
    <span class="n">summaries</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">g</span><span class="p">[</span><span class="s1">&#39;activity_summary&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">lower</span><span class="p">())</span>

    <span class="k">if</span> <span class="p">(</span>
        <span class="s1">&#39;active antagonist&#39;</span> <span class="ow">in</span> <span class="n">summaries</span>
        <span class="ow">and</span> <span class="s1">&#39;active agonist&#39;</span> <span class="ow">in</span> <span class="n">summaries</span>
    <span class="p">):</span>
        <span class="n">allowed_cids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cid</span><span class="p">)</span>

<span class="n">allowed_cids</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">allowed_cids</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;allowed exceptions: </span><span class="si">{</span><span class="n">allowed_cids</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="c1"># Optional inspection</span>
<span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;cid&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">allowed_cids</span><span class="p">)]</span> \
  <span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s1">&#39;cid&#39;</span><span class="p">,</span> <span class="s1">&#39;sid&#39;</span><span class="p">])</span> \
  <span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">20</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>allowed exceptions: {np.int64(10868), np.int64(54454)}
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>pc_result_tag</th>
      <th>sid</th>
      <th>cid</th>
      <th>PUBCHEM_EXT_DATASOURCE_SMILES</th>
      <th>activity_outcome</th>
      <th>activity_score</th>
      <th>activity_url</th>
      <th>assay_data_comment</th>
      <th>activity_summary</th>
      <th>antagonist_activity</th>
      <th>antagonist_potency</th>
      <th>antagonist_efficacy</th>
      <th>viability_activity</th>
      <th>viability_potency</th>
      <th>viability_efficacy</th>
      <th>sample_source</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>6644</th>
      <td>6645</td>
      <td>144210198</td>
      <td>10868</td>
      <td>C1CCC(CC1)N=C=NC2CCCCC2</td>
      <td>Inconclusive</td>
      <td>20.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>active agonist</td>
      <td>active agonist</td>
      <td>4.19020</td>
      <td>40.5990</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>SIGMA</td>
    </tr>
    <tr>
      <th>10000</th>
      <td>10001</td>
      <td>144213554</td>
      <td>10868</td>
      <td>C1CCC(CC1)N=C=NC2CCCCC2</td>
      <td>Active</td>
      <td>50.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>active antagonist</td>
      <td>active antagonist</td>
      <td>5.04574</td>
      <td>-68.1018</td>
      <td>inconclusive antagonist</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>LightBiologicals</td>
    </tr>
    <tr>
      <th>695</th>
      <td>696</td>
      <td>144204247</td>
      <td>54454</td>
      <td>CCC(C)(C)C(=O)O[C@H]1C[C@H](C=C2[C@H]1[C@H]([C...</td>
      <td>Inconclusive</td>
      <td>20.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>active agonist</td>
      <td>active agonist</td>
      <td>4.35177</td>
      <td>54.7546</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>Prestwick Chemical, Inc.</td>
    </tr>
    <tr>
      <th>7164</th>
      <td>7165</td>
      <td>144210718</td>
      <td>54454</td>
      <td>CCC(C)(C)C(=O)O[C@H]1C[C@H](C=C2[C@H]1[C@H]([C...</td>
      <td>Active</td>
      <td>45.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>active antagonist</td>
      <td>active antagonist</td>
      <td>17.22890</td>
      <td>-54.6818</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>Toronto Research</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="alert alert-block alert-info">
<strong>Deeper Dive Explaination</strong>
<p>
  Go to the DYAD pages for
  <a href="https://pubchem.ncbi.nlm.nih.gov/bioassay/743139#sid=144210198&section=Test-Results" target="_blank" rel="noopener noreferrer">
    AID_743139-SID_144210198
  </a>
  and
  <a href="https://pubchem.ncbi.nlm.nih.gov/bioassay/743139#sid=144213554&section=Test-Results" target="_blank" rel="noopener noreferrer">
    AID_743139-SID_144213554
  </a>,
  both of which are CID 10868. Explain the results of the above output.
</p>
  <div style="
    background-color: #efffff;
    color: #000000;
    padding: 10px;
    border-radius: 4px;
    border: 1px solid #dddddd;
    margin-top: 10px;
  ">
<details>
    <summary>Answer</summary>
<p>Although the assay was designed to detect antagonism, it measures receptor signaling in the presence of a reference agonist. Compounds such as CIDs 10868 and 54454 can both activate the receptor on their own and reduce signaling relative to a stronger agonist, consistent with partial agonism. PubChem therefore records agonist-like signal increases as “active agonist” (often marked inconclusive) and signal suppression as “active antagonist.” Both outcomes indicate real target engagement and should be grouped as Active in our binary classification.</p>
</details>
</div>
</div><div class="alert alert-block alert-info">
<strong>Check your understanding</strong>
<ol class="arabic simple">
<li><p>Examine the portion of the dataframe above. Why should CID 16043 be removed from the analysis?</p></li>
<li><p>Can you explain the sequence of the output of <code class="docutils literal notranslate"><span class="pre">df.loc[idx_conflict,:].head(10)</span></code>?  Is this data being sorted?</p></li>
</ol>
  <div style="
    background-color: #efffff;
    color: #000000;
    padding: 10px;
    border-radius: 4px;
    border: 1px solid #dddddd;
    margin-top: 10px;
  ">
<details>
    <summary>Answer</summary>
<ol class="arabic simple">
<li><p>There were three SIDs with a common CID of 16043 and the results were conflictory (two were active, one was inactive)</p></li>
<li><p>This is not sorting, but a grouping by CID occured because the ‘for loop’ iterated through unique CIDs in the order of the idx_conflict list.  When a CID with conflicting activity was found, all rows with that CID were then added to the idx_conflict list.  So index 5 was the first position (CID 16043), then index 5953 and 6847 as they are also CID 16043, now we go to the next index position in the idx_conflict list, which is 49 and there are two CIDs with that value,and then to the next, which is index position 63, and do this for the length of the idx_conflict list.</p></li>
</ol>
</details>
</div>
</div><div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Drop rows with conflicting data, except for allowed biological exceptions</span>
<span class="n">idx_conflict_fixed</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">[</span>
    <span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">idx_conflict</span><span class="p">)</span> <span class="o">&amp;</span>
    <span class="o">~</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;cid&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">allowed_cids</span><span class="p">)</span>
<span class="p">]</span>

<span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">idx_conflict_fixed</span><span class="p">)</span>
<span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>pc_result_tag</th>
      <th>sid</th>
      <th>cid</th>
      <th>PUBCHEM_EXT_DATASOURCE_SMILES</th>
      <th>activity_outcome</th>
      <th>activity_score</th>
      <th>activity_url</th>
      <th>assay_data_comment</th>
      <th>activity_summary</th>
      <th>antagonist_activity</th>
      <th>antagonist_potency</th>
      <th>antagonist_efficacy</th>
      <th>viability_activity</th>
      <th>viability_potency</th>
      <th>viability_efficacy</th>
      <th>sample_source</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>1</td>
      <td>144203552</td>
      <td>12850184</td>
      <td>C(C(=O)[C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O.C(...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2</td>
      <td>144203553</td>
      <td>89753</td>
      <td>C([C@H]([C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O)O...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>2</th>
      <td>3</td>
      <td>144203554</td>
      <td>9403</td>
      <td>C[C@]12CC[C@H]3[C@H]([C@@H]1CC[C@@H]2OC(=O)CCC...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4</td>
      <td>144203555</td>
      <td>13218779</td>
      <td>C[C@@]12CC[C@@H](C1(C)C)C[C@H]2OC(=O)CSC#N</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>9</th>
      <td>10</td>
      <td>144203561</td>
      <td>637566</td>
      <td>CC(=CCC/C(=C/CO)/C)C</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>NCI</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># After removing the conflicting data, we can check the counts of each activity summary again</span>
<span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;activity_summary&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>pc_result_tag</th>
      <th>sid</th>
      <th>cid</th>
      <th>PUBCHEM_EXT_DATASOURCE_SMILES</th>
      <th>activity_outcome</th>
      <th>activity_score</th>
      <th>activity_url</th>
      <th>assay_data_comment</th>
      <th>antagonist_activity</th>
      <th>antagonist_potency</th>
      <th>antagonist_efficacy</th>
      <th>viability_activity</th>
      <th>viability_potency</th>
      <th>viability_efficacy</th>
      <th>sample_source</th>
    </tr>
    <tr>
      <th>activity_summary</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>active agonist</th>
      <td>539</td>
      <td>539</td>
      <td>539</td>
      <td>539</td>
      <td>539</td>
      <td>539</td>
      <td>0</td>
      <td>0</td>
      <td>539</td>
      <td>539</td>
      <td>539</td>
      <td>539</td>
      <td>58</td>
      <td>519</td>
      <td>539</td>
    </tr>
    <tr>
      <th>active antagonist</th>
      <td>344</td>
      <td>344</td>
      <td>344</td>
      <td>344</td>
      <td>344</td>
      <td>344</td>
      <td>0</td>
      <td>0</td>
      <td>344</td>
      <td>343</td>
      <td>344</td>
      <td>344</td>
      <td>108</td>
      <td>327</td>
      <td>344</td>
    </tr>
    <tr>
      <th>inactive</th>
      <td>7386</td>
      <td>7386</td>
      <td>7386</td>
      <td>7386</td>
      <td>7386</td>
      <td>7386</td>
      <td>0</td>
      <td>0</td>
      <td>7386</td>
      <td>0</td>
      <td>7386</td>
      <td>7386</td>
      <td>318</td>
      <td>7275</td>
      <td>7386</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># identify Compounds IDs and Substance IDs. </span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;total number of Substance IDs=&#39;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;sid&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;total number of Compound IDs= &#39;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;cid&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>total number of Substance IDs= 8269
total number of Compound IDs=  6793
</pre></div>
</div>
</div>
</div>
</section>
<section id="remove-redundant-data">
<h2>4.4 Remove redundant data<a class="headerlink" href="#remove-redundant-data" title="Link to this heading">#</a></h2>
<p>The code cells in section 4.3. do not remove compounds tested multiple times if the testing results are consistent (e.g., active agonist in all samples (substances)).  The rows corresponding to these compounds are redundant, so we want remove them except for only one row for each compound. To illustrate this, let’s create a sorted dataframe to see some of the redundancies.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sorted_df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;cid&#39;</span><span class="p">,</span><span class="s1">&#39;sid&#39;</span><span class="p">,</span><span class="s1">&#39;activity_summary&#39;</span><span class="p">],</span>
                           <span class="n">ascending</span><span class="o">=</span><span class="p">[</span><span class="kc">True</span><span class="p">,</span><span class="kc">True</span><span class="p">,</span><span class="kc">True</span><span class="p">])</span>
<span class="n">sorted_df</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>pc_result_tag</th>
      <th>sid</th>
      <th>cid</th>
      <th>PUBCHEM_EXT_DATASOURCE_SMILES</th>
      <th>activity_outcome</th>
      <th>activity_score</th>
      <th>activity_url</th>
      <th>assay_data_comment</th>
      <th>activity_summary</th>
      <th>antagonist_activity</th>
      <th>antagonist_potency</th>
      <th>antagonist_efficacy</th>
      <th>viability_activity</th>
      <th>viability_potency</th>
      <th>viability_efficacy</th>
      <th>sample_source</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>8467</th>
      <td>8468</td>
      <td>144212021</td>
      <td>4</td>
      <td>CC(CN)O</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>Enamine</td>
    </tr>
    <tr>
      <th>6110</th>
      <td>6111</td>
      <td>144209664</td>
      <td>11</td>
      <td>C(CCl)Cl</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>SIGMA</td>
    </tr>
    <tr>
      <th>4955</th>
      <td>4956</td>
      <td>144208509</td>
      <td>13</td>
      <td>C1=CC(=C(C=C1Cl)Cl)Cl</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>SIGMA</td>
    </tr>
    <tr>
      <th>7326</th>
      <td>7327</td>
      <td>144210880</td>
      <td>13</td>
      <td>C1=CC(=C(C=C1Cl)Cl)Cl</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>LightBiologicals</td>
    </tr>
    <tr>
      <th>7915</th>
      <td>7916</td>
      <td>144211469</td>
      <td>33</td>
      <td>C(C=O)Cl</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>LightBiologicals</td>
    </tr>
    <tr>
      <th>6809</th>
      <td>6810</td>
      <td>144210363</td>
      <td>34</td>
      <td>C(CCl)O</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>Enamine</td>
    </tr>
    <tr>
      <th>15</th>
      <td>16</td>
      <td>144203567</td>
      <td>51</td>
      <td>C(CC(=O)O)C(=O)C(=O)O</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>4562</th>
      <td>4563</td>
      <td>144208116</td>
      <td>51</td>
      <td>C(CC(=O)O)C(=O)C(=O)O</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>SIGMA</td>
    </tr>
    <tr>
      <th>5533</th>
      <td>5534</td>
      <td>144209087</td>
      <td>66</td>
      <td>C1=CC(=C(C=C1Cl)O)Cl</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>SIGMA</td>
    </tr>
    <tr>
      <th>9816</th>
      <td>9817</td>
      <td>144213370</td>
      <td>66</td>
      <td>C1=CC(=C(C=C1Cl)O)Cl</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>SIGMA</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="alert alert-block alert-warning">
Explore the above displayed sorted dataframe. Identify any CIDs that are repeated, and if they have different <b>SID</b> or <b>activity_outcome</b> values.<p>We will use the <code class="docutils literal notranslate"><span class="pre">drop_duplicates()</span></code> method in pandas to remove any duplicate rows in our dataframe. We can use the <code class="docutils literal notranslate"><span class="pre">subset</span></code> parameter to signify which column to consider when identifying duplicates. In our case, we need unique identifiers to get structural data, so we use <strong>‘cid’</strong>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_clean</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="s1">&#39;cid&#39;</span><span class="p">)</span>  <span class="c1"># remove duplicate rows except for the first occurring row.</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;total number of Substance IDs=&#39;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">df_clean</span><span class="p">[</span><span class="s1">&#39;sid&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;total number of Compound IDs= &#39;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">df_clean</span><span class="p">[</span><span class="s1">&#39;cid&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>total number of Substance IDs= 6793
total number of Compound IDs=  6793
</pre></div>
</div>
</div>
</div>
<section id="save-as-aid743139-clean-cid">
<h3>Save as AID743139_clean_CID…<a class="headerlink" href="#save-as-aid743139-clean-cid" title="Link to this heading">#</a></h3>
<p>Save the dataframe in your curated data folder with date stamp and version number using <code class="docutils literal notranslate"><span class="pre">prefix=&quot;AID743139_clean_CID&quot;,</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Complete the code to save your work as a versioned filename in the folder</span>
<span class="c1"># 10_SupervisedMLdata/AID743139/curated/</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">cinf26pk.core</span><span class="w"> </span><span class="kn">import</span> <span class="n">filenames</span>
<span class="c1"># generate path</span>
<span class="n">PROJECT_ROOT</span> <span class="o">=</span> <span class="n">Path</span><span class="o">.</span><span class="n">cwd</span><span class="p">()</span> <span class="c1"># project root (10_SupervisedML)</span>
<span class="n">CUR</span> <span class="o">=</span> <span class="n">PROJECT_ROOT</span> <span class="o">/</span> <span class="s2">&quot;data&quot;</span> <span class="o">/</span> <span class="s2">&quot;AID743139&quot;</span> <span class="o">/</span> <span class="s2">&quot;curated&quot;</span>
<span class="n">CUR</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Create deterministic filename with date stamp</span>
<span class="n">fname</span> <span class="o">=</span> <span class="n">filenames</span><span class="o">.</span><span class="n">make_filename</span><span class="p">(</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;AID743139_clean_CID&quot;</span><span class="p">,</span>
    <span class="n">ext</span><span class="o">=</span><span class="s2">&quot;csv&quot;</span><span class="p">,</span>
    <span class="n">folder</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">CUR</span><span class="p">)</span>
<span class="p">)</span>

<span class="c1"># Save raw dataset</span>
<span class="n">df_clean</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">CUR</span><span class="o">/</span><span class="n">fname</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[Saved] Curated PubChem data → </span><span class="si">{</span><span class="n">fname</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[Saved] Curated PubChem data → /home/rebelford/jupyterbooks/cinf26book/content/modules/10_SupervisedML/data/AID743139/curated/AID743139_clean_CID_20260118_v1.csv
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#write your code here to open the file as a dataframe called df </span>
<span class="c1">#double check the the number of CIDs and SIDs has not changedand check that inconclusives and duplicates have been removed</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">CUR</span><span class="o">/</span><span class="s2">&quot;AID743139_clean_CID_20260118_v1.csv&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;total number of Substance IDs=&#39;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;sid&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;total number of Compound IDs= &#39;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;cid&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()))</span>
<span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>total number of Substance IDs= 6793
total number of Compound IDs=  6793
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>pc_result_tag</th>
      <th>sid</th>
      <th>cid</th>
      <th>PUBCHEM_EXT_DATASOURCE_SMILES</th>
      <th>activity_outcome</th>
      <th>activity_score</th>
      <th>activity_url</th>
      <th>assay_data_comment</th>
      <th>activity_summary</th>
      <th>antagonist_activity</th>
      <th>antagonist_potency</th>
      <th>antagonist_efficacy</th>
      <th>viability_activity</th>
      <th>viability_potency</th>
      <th>viability_efficacy</th>
      <th>sample_source</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>1</td>
      <td>144203552</td>
      <td>12850184</td>
      <td>C(C(=O)[C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O.C(...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2</td>
      <td>144203553</td>
      <td>89753</td>
      <td>C([C@H]([C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O)O...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>2</th>
      <td>3</td>
      <td>144203554</td>
      <td>9403</td>
      <td>C[C@]12CC[C@H]3[C@H]([C@@H]1CC[C@@H]2OC(=O)CCC...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4</td>
      <td>144203555</td>
      <td>13218779</td>
      <td>C[C@@]12CC[C@@H](C1(C)C)C[C@H]2OC(=O)CSC#N</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>NCI</td>
    </tr>
    <tr>
      <th>4</th>
      <td>10</td>
      <td>144203561</td>
      <td>637566</td>
      <td>CC(=CCC/C(=C/CO)/C)C</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>NCI</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Note from the above that cids are now integers, and the number of CIDs equal the number of SIDs</p>
</section>
</section>
<section id="adding-numeric-activity-classes">
<h2>4.5 Adding “numeric” activity classes<a class="headerlink" href="#adding-numeric-activity-classes" title="Link to this heading">#</a></h2>
<p>In general, machine learning algorithms require both inputs and outputs to be in numerical form.  In our case the input we will be using are molecular structure. We have been working to clean our data to ensure we have valid CIDs that we can use to retrieve SMILES strings and subsequently calculate binary fingerprints as input features.  However. the output of biological activity is currently stored as text labels as active or inactive. To make this data suitable for modeling, we will create a new column called <strong>‘activity’</strong> that assigns these labels as numeric values:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">1</span></code> for actives (including both active agonists and antagonists)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">0</span></code> for inactives</p></li>
</ul>
<p>We combine the two types of active compounds into a single class because our goal is to train a binary classifier that distinguishes between compounds that interact with the target (active) and those that do not (inactive).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;activity_outcome&#39;</span><span class="p">,</span><span class="s1">&#39;activity_summary&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th>pc_result_tag</th>
      <th>sid</th>
      <th>cid</th>
      <th>PUBCHEM_EXT_DATASOURCE_SMILES</th>
      <th>activity_score</th>
      <th>activity_url</th>
      <th>assay_data_comment</th>
      <th>antagonist_activity</th>
      <th>antagonist_potency</th>
      <th>antagonist_efficacy</th>
      <th>viability_activity</th>
      <th>viability_potency</th>
      <th>viability_efficacy</th>
      <th>sample_source</th>
    </tr>
    <tr>
      <th>activity_outcome</th>
      <th>activity_summary</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Active</th>
      <th>active antagonist</th>
      <td>290</td>
      <td>290</td>
      <td>290</td>
      <td>290</td>
      <td>290</td>
      <td>0</td>
      <td>0</td>
      <td>290</td>
      <td>289</td>
      <td>290</td>
      <td>290</td>
      <td>88</td>
      <td>275</td>
      <td>290</td>
    </tr>
    <tr>
      <th>Inactive</th>
      <th>inactive</th>
      <td>6050</td>
      <td>6050</td>
      <td>6050</td>
      <td>6050</td>
      <td>6050</td>
      <td>0</td>
      <td>0</td>
      <td>6050</td>
      <td>0</td>
      <td>6050</td>
      <td>6050</td>
      <td>269</td>
      <td>5964</td>
      <td>6050</td>
    </tr>
    <tr>
      <th>Inconclusive</th>
      <th>active agonist</th>
      <td>453</td>
      <td>453</td>
      <td>453</td>
      <td>453</td>
      <td>453</td>
      <td>0</td>
      <td>0</td>
      <td>453</td>
      <td>453</td>
      <td>453</td>
      <td>453</td>
      <td>44</td>
      <td>434</td>
      <td>453</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;activity&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span> <span class="mi">0</span> <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="s1">&#39;inactive&#39;</span> <span class="k">else</span> <span class="mi">1</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;activity_summary&#39;</span><span class="p">]</span> <span class="p">]</span>
<span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span> <span class="c1">#new column is added to end of column list, so scroll right in the resulting cell output.</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>pc_result_tag</th>
      <th>sid</th>
      <th>cid</th>
      <th>PUBCHEM_EXT_DATASOURCE_SMILES</th>
      <th>activity_outcome</th>
      <th>activity_score</th>
      <th>activity_url</th>
      <th>assay_data_comment</th>
      <th>activity_summary</th>
      <th>antagonist_activity</th>
      <th>antagonist_potency</th>
      <th>antagonist_efficacy</th>
      <th>viability_activity</th>
      <th>viability_potency</th>
      <th>viability_efficacy</th>
      <th>sample_source</th>
      <th>activity</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>1</td>
      <td>144203552</td>
      <td>12850184</td>
      <td>C(C(=O)[C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O.C(...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>NCI</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2</td>
      <td>144203553</td>
      <td>89753</td>
      <td>C([C@H]([C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O)O...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>NCI</td>
      <td>0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>3</td>
      <td>144203554</td>
      <td>9403</td>
      <td>C[C@]12CC[C@H]3[C@H]([C@@H]1CC[C@@H]2OC(=O)CCC...</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>NCI</td>
      <td>0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4</td>
      <td>144203555</td>
      <td>13218779</td>
      <td>C[C@@]12CC[C@@H](C1(C)C)C[C@H]2OC(=O)CSC#N</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>NCI</td>
      <td>0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>10</td>
      <td>144203561</td>
      <td>637566</td>
      <td>CC(=CCC/C(=C/CO)/C)C</td>
      <td>Inactive</td>
      <td>0.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>inactive</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>inactive</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>NCI</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Double-check the count of active/inactive compounds.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;activity_summary&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>pc_result_tag</th>
      <th>sid</th>
      <th>cid</th>
      <th>PUBCHEM_EXT_DATASOURCE_SMILES</th>
      <th>activity_outcome</th>
      <th>activity_score</th>
      <th>activity_url</th>
      <th>assay_data_comment</th>
      <th>antagonist_activity</th>
      <th>antagonist_potency</th>
      <th>antagonist_efficacy</th>
      <th>viability_activity</th>
      <th>viability_potency</th>
      <th>viability_efficacy</th>
      <th>sample_source</th>
      <th>activity</th>
    </tr>
    <tr>
      <th>activity_summary</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>active agonist</th>
      <td>453</td>
      <td>453</td>
      <td>453</td>
      <td>453</td>
      <td>453</td>
      <td>453</td>
      <td>0</td>
      <td>0</td>
      <td>453</td>
      <td>453</td>
      <td>453</td>
      <td>453</td>
      <td>44</td>
      <td>434</td>
      <td>453</td>
      <td>453</td>
    </tr>
    <tr>
      <th>active antagonist</th>
      <td>290</td>
      <td>290</td>
      <td>290</td>
      <td>290</td>
      <td>290</td>
      <td>290</td>
      <td>0</td>
      <td>0</td>
      <td>290</td>
      <td>289</td>
      <td>290</td>
      <td>290</td>
      <td>88</td>
      <td>275</td>
      <td>290</td>
      <td>290</td>
    </tr>
    <tr>
      <th>inactive</th>
      <td>6050</td>
      <td>6050</td>
      <td>6050</td>
      <td>6050</td>
      <td>6050</td>
      <td>6050</td>
      <td>0</td>
      <td>0</td>
      <td>6050</td>
      <td>0</td>
      <td>6050</td>
      <td>6050</td>
      <td>269</td>
      <td>5964</td>
      <td>6050</td>
      <td>6050</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;activity&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">count</span><span class="p">()</span> 
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>pc_result_tag</th>
      <th>sid</th>
      <th>cid</th>
      <th>PUBCHEM_EXT_DATASOURCE_SMILES</th>
      <th>activity_outcome</th>
      <th>activity_score</th>
      <th>activity_url</th>
      <th>assay_data_comment</th>
      <th>activity_summary</th>
      <th>antagonist_activity</th>
      <th>antagonist_potency</th>
      <th>antagonist_efficacy</th>
      <th>viability_activity</th>
      <th>viability_potency</th>
      <th>viability_efficacy</th>
      <th>sample_source</th>
    </tr>
    <tr>
      <th>activity</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>6050</td>
      <td>6050</td>
      <td>6050</td>
      <td>6050</td>
      <td>6050</td>
      <td>6050</td>
      <td>0</td>
      <td>0</td>
      <td>6050</td>
      <td>6050</td>
      <td>0</td>
      <td>6050</td>
      <td>6050</td>
      <td>269</td>
      <td>5964</td>
      <td>6050</td>
    </tr>
    <tr>
      <th>1</th>
      <td>743</td>
      <td>743</td>
      <td>743</td>
      <td>743</td>
      <td>743</td>
      <td>743</td>
      <td>0</td>
      <td>0</td>
      <td>743</td>
      <td>743</td>
      <td>742</td>
      <td>743</td>
      <td>743</td>
      <td>132</td>
      <td>709</td>
      <td>743</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="alert alert-block alert-warning">
Do the number of actives in this output equal the active agonist and active antagonist above?</section>
<section id="create-a-smaller-data-frame-that-only-contains-cids-and-activities">
<h2>4.6 Create a smaller data frame that only contains CIDs and activities.<a class="headerlink" href="#create-a-smaller-data-frame-that-only-contains-cids-and-activities" title="Link to this heading">#</a></h2>
<p>Let’s create a smaller data frame that only contains CIDs, activities and the PubChem External Data Sources.  This data frame will be merged with a data frame containing molecular fingerprint information.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_activity</span> <span class="o">=</span> <span class="n">df</span><span class="p">[[</span><span class="s1">&#39;cid&#39;</span><span class="p">,</span><span class="s1">&#39;activity&#39;</span><span class="p">,</span><span class="s1">&#39;PUBCHEM_EXT_DATASOURCE_SMILES&#39;</span><span class="p">]]</span>  <span class="c1">#also included the external datasource SMILES to compare later with PubChem validated CID SMILES</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_activity</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cid</th>
      <th>activity</th>
      <th>PUBCHEM_EXT_DATASOURCE_SMILES</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>12850184</td>
      <td>0</td>
      <td>C(C(=O)[C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O.C(...</td>
    </tr>
    <tr>
      <th>1</th>
      <td>89753</td>
      <td>0</td>
      <td>C([C@H]([C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O)O...</td>
    </tr>
    <tr>
      <th>2</th>
      <td>9403</td>
      <td>0</td>
      <td>C[C@]12CC[C@H]3[C@H]([C@@H]1CC[C@@H]2OC(=O)CCC...</td>
    </tr>
    <tr>
      <th>3</th>
      <td>13218779</td>
      <td>0</td>
      <td>C[C@@]12CC[C@@H](C1(C)C)C[C@H]2OC(=O)CSC#N</td>
    </tr>
    <tr>
      <th>4</th>
      <td>637566</td>
      <td>0</td>
      <td>CC(=CCC/C(=C/CO)/C)C</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_activity</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(6793, 3)
</pre></div>
</div>
</div>
</div>
<p>While we could have used the SMILES that were already in our dataframe, the label was PUBCHEM_EXT_DATASOURCE_SMILES indicating that the SMILES may have been provided by the submitter, and not cannonical.  To ensure we have quality data, we will retrieve SMILES from PubChem.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;CIDs are stored in original df dataframe as&quot;</span><span class="p">,</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;cid&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">dtype</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;CIDs are stored in df_activity dataframe as&quot;</span><span class="p">,</span> <span class="n">df_activity</span><span class="p">[</span><span class="s1">&#39;cid&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">dtype</span><span class="p">)</span>
<span class="n">cids</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">cid</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">cids</span><span class="p">[:</span><span class="mi">10</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;CIDs are now stored as&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="p">(</span><span class="n">cids</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;CIDs is type </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">cids</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>CIDs are stored in original df dataframe as int64
CIDs are stored in df_activity dataframe as int64
[12850184, 89753, 9403, 13218779, 637566, 4766, 3080, 7048801, 51, 66435]
CIDs are now stored as &lt;class &#39;int&#39;&gt;
CIDs is type &lt;class &#39;list&#39;&gt;
</pre></div>
</div>
</div>
</div>
<section id="save-as-aid743139-activity-cid">
<h3>4.6.1 Save as AID743139_activity_CID<a class="headerlink" href="#save-as-aid743139-activity-cid" title="Link to this heading">#</a></h3>
<p>Save the dataframe in your curated data folder with date stamp and version number using <code class="docutils literal notranslate"><span class="pre">prefix=&quot;AID743139_activity_CID&quot;,</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#write code here</span>
<span class="c1"># Complete the code to save your work as a versioned filename in the folder</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">cinf26pk.core</span><span class="w"> </span><span class="kn">import</span> <span class="n">filenames</span>
<span class="c1"># generate path</span>
<span class="n">PROJECT_ROOT</span> <span class="o">=</span> <span class="n">Path</span><span class="o">.</span><span class="n">cwd</span><span class="p">()</span> <span class="c1"># project root (10_SupervisedML)</span>
<span class="n">CUR</span> <span class="o">=</span> <span class="n">PROJECT_ROOT</span> <span class="o">/</span> <span class="s2">&quot;data&quot;</span> <span class="o">/</span> <span class="s2">&quot;AID743139&quot;</span> <span class="o">/</span> <span class="s2">&quot;curated&quot;</span>
<span class="n">CUR</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Create deterministic filename with date stamp</span>
<span class="n">fname</span> <span class="o">=</span> <span class="n">filenames</span><span class="o">.</span><span class="n">make_filename</span><span class="p">(</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;AID743139_Activity_CID&quot;</span><span class="p">,</span>
    <span class="n">ext</span><span class="o">=</span><span class="s2">&quot;csv&quot;</span><span class="p">,</span>
    <span class="n">folder</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">CUR</span><span class="p">)</span>
<span class="p">)</span>

<span class="c1"># Save raw dataset</span>
<span class="n">df_activity</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">CUR</span><span class="o">/</span><span class="n">fname</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[Saved] Curated PubChem data → </span><span class="si">{</span><span class="n">fname</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[Saved] Curated PubChem data → /home/rebelford/jupyterbooks/cinf26book/content/modules/10_SupervisedML/data/AID743139/curated/AID743139_Activity_CID_20260118_v1.csv
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="acquire-standardized-pubchem-smiles">
<h1>5 Acquire standardized PubChem SMILES<a class="headerlink" href="#acquire-standardized-pubchem-smiles" title="Link to this heading">#</a></h1>
<p>Now that we have a list of CIDs as integers, we can retrive the SMILES from PubChem using the PUG REST API. We will use these to generate RDKit fingerprints.</p>
<p>Once we have retrieved the SMILES we will store them and their respective CIDs into a new <code class="docutils literal notranslate"><span class="pre">df_smiles</span></code> dataframe.</p>
<div class="alert alert-block alert-success">
<strong>Code Dive</strong>
<p>The following code uses several new functions in the cinf26pk.py package, along with the tqdm package.  You can insert the cinf26pk functions directly into the code cell if you do not have access to the package (after removing the import statements)
</p>
  <div style="
    background-color: #efffff;
    color: #000000;
    padding: 10px;
    border-radius: 4px;
    border: 1px solid #dddddd;
    margin-top: 10px;
  ">
<details>
    <summary>chunk_list() - a function of the cinf26.core.utils module</summary>
<pre><code>      
1)def chunk_list(lst: List[Any], size: int = 100):
    """
    Yield successive fixed-size chunks from a list.
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Parameters
----------
lst : list
    List to split.
size : int, default=100
    Chunk size.

Yields
------
list
    List chunk.
&quot;&quot;&quot;
for i in range(0, len(lst), size):
    yield lst[i:i + size]
</pre></div>
</div>
<p></code></pre></p>
</details>
</div>
  <div style="
    background-color: #efffff;
    color: #000000;
    padding: 10px;
    border-radius: 4px;
    border: 1px solid #dddddd;
    margin-top: 10px;
  ">
<details>
    <summary>pubchem_get() - a function of the cinf26.pubchem.http module</summary>
<pre><code>
def pubchem_get(url, retries=DEFAULT_RETRIES, timeout=DEFAULT_TIMEOUT,
                base_delay=2, return_json=True):
    """
    Safely performs a GET request to a PubChem PUG-REST endpoint with retry/backoff.
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Parameters
----------
url : str
    Full URL for the GET request.
retries : int, default=3
    Number of retry attempts.
timeout : int, default=30
    Timeout (seconds) for request.
base_delay : int, default=2
    Backoff base delay.
return_json : bool, default=True
    Return JSON or raw text.

Returns
-------
dict or str or None
    Parsed response, or None if all retries fail.
&quot;&quot;&quot;  
for attempt in range(1, retries + 1):
    try:
        r = requests.get(url, timeout=timeout)
        r.raise_for_status()
        return r.json() if return_json else r.text

    except Exception as e:
        print(f&quot;[Warning] GET attempt {attempt}/{retries} failed.&quot;)
        print(f&quot;         URL: {url}&quot;)
        print(f&quot;         Error: {type(e).__name__}: {e}&quot;)
        time.sleep(base_delay * attempt)

print(f&quot;[Error] All GET attempts failed for URL: {url}&quot;)
return None
</pre></div>
</div>
<p></code></pre></p>
</details>
</div>
  <div style="
    background-color: #efffff;
    color: #000000;
    padding: 10px;
    border-radius: 4px;
    border: 1px solid #dddddd;
    margin-top: 10px;
  ">
<details>
    <summary>tqdm</summary>
<p>
<strong>tqdm</strong> is a Python package that provides fast, extensible progress bars for loops and iterable-based operations. It is commonly used in data science and machine learning workflows to give real-time feedback on long-running tasks such as file I/O, model training, or batch processing, without requiring changes to the underlying logic of the code. tqdm integrates cleanly with standard Python iterables, Pandas operations, and Jupyter notebooks.
</p>
</p>To install activate your env and use conda-forge
<pre><code>conda install -c conda-forge tqdm</code></pre>
<p>
</details>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">cinf26pk.core</span><span class="w"> </span><span class="kn">import</span> <span class="n">utils</span><span class="p">,</span> <span class="n">filenames</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">cinf26pk.pubchem</span><span class="w"> </span><span class="kn">import</span> <span class="n">http</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">io</span><span class="w"> </span><span class="kn">import</span> <span class="n">StringIO</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">tqdm</span><span class="w"> </span><span class="kn">import</span> <span class="n">tqdm</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">time</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="c1"># 1. If needed, Reload df_activity from your saved CSV </span>
<span class="n">df_activity</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;data/AID743139/curated/AID743139_Activity_CID_20260118_v1.csv&quot;</span><span class="p">)</span>
<span class="c1"># 2. Extract CIDs as integers</span>
<span class="n">cids</span> <span class="o">=</span> <span class="n">df_activity</span><span class="o">.</span><span class="n">cid</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;# CIDs:&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">cids</span><span class="p">))</span>

<span class="c1"># 3. Retrieve standardized PubChem SMILES</span>
<span class="n">chunk_size</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">sleep_time</span> <span class="o">=</span> <span class="mf">0.2</span>   <span class="c1"># 5 calls per second</span>

<span class="n">base_url</span> <span class="o">=</span> <span class="s2">&quot;https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/cid&quot;</span>

<span class="n">df_chunks</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">failed_chunks</span> <span class="o">=</span> <span class="p">[]</span>

<span class="n">chunks</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">utils</span><span class="o">.</span><span class="n">chunk_list</span><span class="p">(</span><span class="n">cids</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">chunk_size</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;# Total chunks:&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">chunks</span><span class="p">))</span>

<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">tqdm</span><span class="p">(</span><span class="n">chunks</span><span class="p">)):</span>
    
    <span class="n">cid_str</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">chunk</span><span class="p">)</span>
    <span class="n">url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">base_url</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">cid_str</span><span class="si">}</span><span class="s2">/property/SMILES/TXT&quot;</span>

    <span class="n">text</span> <span class="o">=</span> <span class="n">http</span><span class="o">.</span><span class="n">pubchem_get</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="n">return_json</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="n">text</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[Warning] Failed chunk </span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">failed_chunks</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="k">continue</span>
    
    <span class="c1"># Convert to DataFrame</span>
    <span class="n">df_chunk</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">StringIO</span><span class="p">(</span><span class="n">text</span><span class="p">),</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;smiles&#39;</span><span class="p">])</span>
    <span class="n">df_chunks</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">df_chunk</span><span class="p">)</span>
    
    <span class="c1"># Rate limit compliance</span>
    <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="n">sleep_time</span><span class="p">)</span>

<span class="c1"># 4. Combine all SMILES results</span>
<span class="n">df_smiles</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">df_chunks</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">df_smiles</span><span class="p">[</span><span class="s2">&quot;cid&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cids</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Preview of combined SMILES:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df_smiles</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Failed chunks: </span><span class="si">{</span><span class="n">failed_chunks</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># 5. Save to data/ with versioning</span>
<span class="c1"># generate path</span>
<span class="n">PROJECT_ROOT</span> <span class="o">=</span> <span class="n">Path</span><span class="o">.</span><span class="n">cwd</span><span class="p">()</span> <span class="c1"># project root (10_SupervisedML)</span>
<span class="n">CUR</span> <span class="o">=</span> <span class="n">PROJECT_ROOT</span> <span class="o">/</span> <span class="s2">&quot;data&quot;</span> <span class="o">/</span> <span class="s2">&quot;AID743139&quot;</span> <span class="o">/</span> <span class="s2">&quot;curated&quot;</span>
<span class="n">CUR</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Create deterministic filename with date stamp</span>
<span class="n">fname</span> <span class="o">=</span> <span class="n">filenames</span><span class="o">.</span><span class="n">make_filename</span><span class="p">(</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;AID743139_pcSMILES_act&quot;</span><span class="p">,</span>
    <span class="n">ext</span><span class="o">=</span><span class="s2">&quot;csv&quot;</span><span class="p">,</span>
    <span class="n">folder</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">CUR</span><span class="p">)</span>
<span class="p">)</span>

<span class="c1"># Save raw dataset</span>
<span class="n">df_smiles</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">CUR</span><span class="o">/</span><span class="n">fname</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[Saved] Curated PubChem data → </span><span class="si">{</span><span class="n">fname</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span># CIDs: 6793
# Total chunks: 68
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|███████████████████████████████████████████████████████████████████████████████████| 68/68 [05:56&lt;00:00,  5.25s/it]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Preview of combined SMILES:
                                              smiles       cid
0  C(C(=O)[C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O.C(...  12850184
1  C([C@H]([C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O)O...     89753
2  C[C@]12CC[C@H]3[C@H]([C@@H]1CC[C@@H]2OC(=O)CCC...      9403
3         C[C@@]12CC[C@@H](C1(C)C)C[C@H]2OC(=O)CSC#N  13218779
4                               CC(=CCC/C(=C/CO)/C)C    637566

Failed chunks: []
[Saved] Curated PubChem data → /home/rebelford/jupyterbooks/cinf26book/content/modules/10_SupervisedML/data/AID743139/curated/AID743139_pcSMILES_act_20260118_v1.csv
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_smiles</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>smiles</th>
      <th>cid</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>C(C(=O)[C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O.C(...</td>
      <td>12850184</td>
    </tr>
    <tr>
      <th>1</th>
      <td>C([C@H]([C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O)O...</td>
      <td>89753</td>
    </tr>
    <tr>
      <th>2</th>
      <td>C[C@]12CC[C@H]3[C@H]([C@@H]1CC[C@@H]2OC(=O)CCC...</td>
      <td>9403</td>
    </tr>
    <tr>
      <th>3</th>
      <td>C[C@@]12CC[C@@H](C1(C)C)C[C@H]2OC(=O)CSC#N</td>
      <td>13218779</td>
    </tr>
    <tr>
      <th>4</th>
      <td>CC(=CCC/C(=C/CO)/C)C</td>
      <td>637566</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<section id="deal-with-salts-and-mixtures">
<h2>5.1  Deal with salts and mixtures<a class="headerlink" href="#deal-with-salts-and-mixtures" title="Link to this heading">#</a></h2>
<p>We will be generating fingerprints from the smiles strings.  To ensure the quality of the fingerprints, we will clean the SMILES strings to remove salts and keep only the largest fragment of each compound. This is important because salts and mixtures can introduce noise and variability in the fingerprint data, which can negatively impact model performance.</p>
<p><code class="docutils literal notranslate"><span class="pre">SaltRemover</span></code> is a utility module within <strong>RDKit</strong> that remove counterions and small disconnected fragments from molecular structures. Many chemical records represent compounds in their salted or formulated forms (e.g., hydrochloride salts, sodium salts, solvates). While chemically valid, these extra components can interfere with structure-based analysis, fingerprint generation, and machine-learning workflows. The purpose of <code class="docutils literal notranslate"><span class="pre">SaltRemover</span></code> is therefore not to change the underlying chemical identity of a compound, but to isolate the core molecular structure that is typically responsible for biological activity.</p>
<p>In the following code, each SMILES string is first converted into an RDKit molecule object using <code class="docutils literal notranslate"><span class="pre">Chem.MolFromSmiles</span></code>. The <code class="docutils literal notranslate"><span class="pre">SaltRemover</span></code> is then applied via <code class="docutils literal notranslate"><span class="pre">StripMol</span></code>, which removes known salt fragments while ensuring that at least one fragment remains (<code class="docutils literal notranslate"><span class="pre">dontRemoveEverything=True</span></code>). After salt stripping, the molecule may still consist of multiple disconnected fragments (for example, residual solvents or small ions). To resolve this, the code explicitly extracts all fragments and retains only the largest fragment by atom count, under the assumption that this fragment represents the biologically relevant parent compound. Finally, the cleaned molecule is converted back into a SMILES string using <code class="docutils literal notranslate"><span class="pre">Chem.MolToSmiles</span></code>, and the result is stored as <code class="docutils literal notranslate"><span class="pre">clean_smiles</span></code>. The net effect is a standardized, salt-free representation of each compound suitable for fingerprint generation, comparison, and machine-learning analysis.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># reload the file you just saved and print the head to ensure it was correctly saved</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="n">df_smiles</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;data/AID743139/curated/AID743139_pcSMILES_act_20260118_v1.csv&quot;</span><span class="p">)</span>
<span class="n">df_smiles</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>smiles</th>
      <th>cid</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>C(C(=O)[C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O.C(...</td>
      <td>12850184</td>
    </tr>
    <tr>
      <th>1</th>
      <td>C([C@H]([C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O)O...</td>
      <td>89753</td>
    </tr>
    <tr>
      <th>2</th>
      <td>C[C@]12CC[C@H]3[C@H]([C@@H]1CC[C@@H]2OC(=O)CCC...</td>
      <td>9403</td>
    </tr>
    <tr>
      <th>3</th>
      <td>C[C@@]12CC[C@@H](C1(C)C)C[C@H]2OC(=O)CSC#N</td>
      <td>13218779</td>
    </tr>
    <tr>
      <th>4</th>
      <td>CC(=CCC/C(=C/CO)/C)C</td>
      <td>637566</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="alert alert-block alert-success">
<strong>Code Dive</strong>
<p>This line of code in the following code cell.
</p>
<pre><code>df_smiles['clean_smiles'] = df_smiles['smiles'].apply(clean_smiles)
</pre></code>
  <div style="
    background-color: #efffff;
    color: #000000;
    padding: 10px;
    border-radius: 4px;
    border: 1px solid #dddddd;
    margin-top: 10px;
  ">
<details>
    <summary>Explanation</summary>
In the above code we created a python function called clean_smiles().  The apply method allowed us to apply it to the "smiles" series and generate a new "clean_smiles" series within the DataFrame, which is a column of salt-free SMILES. Although it behaves like a loop, it is written in a Pandas-native way that preserves index alignment and avoids manual iteration. 
</details>
</div>
</div>
<p>To identify how many smiles needed to be cleaned you could compare the smiles in the ‘smiles’  column with those in the ‘clean_smiles’ column but due to the noncanonical nature of smiles those generated by RDKit may differ than those obtained by PubChem, even if the structure has undergone on modification.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">rdkit</span><span class="w"> </span><span class="kn">import</span> <span class="n">Chem</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">rdkit.Chem</span><span class="w"> </span><span class="kn">import</span> <span class="n">SaltRemover</span>

<span class="n">remover</span> <span class="o">=</span> <span class="n">SaltRemover</span><span class="o">.</span><span class="n">SaltRemover</span><span class="p">()</span>

<span class="k">def</span><span class="w"> </span><span class="nf">clean_smiles</span><span class="p">(</span><span class="n">sm</span><span class="p">):</span>
    <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromSmiles</span><span class="p">(</span><span class="n">sm</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">mol</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="n">mol</span> <span class="o">=</span> <span class="n">remover</span><span class="o">.</span><span class="n">StripMol</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">dontRemoveEverything</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># keep only largest fragment</span>
    <span class="n">frags</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">GetMolFrags</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">asMols</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">mol</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">frags</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">m</span><span class="p">:</span> <span class="n">m</span><span class="o">.</span><span class="n">GetNumAtoms</span><span class="p">())</span>

    <span class="k">return</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolToSmiles</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>

<span class="n">df_smiles</span><span class="p">[</span><span class="s1">&#39;clean_smiles&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_smiles</span><span class="p">[</span><span class="s1">&#39;smiles&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">clean_smiles</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">df_smiles</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df_smiles</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[08:50:44] WARNING: not removing hydrogen atom without neighbors
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>                                              smiles       cid  \
0  C(C(=O)[C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O.C(...  12850184   
1  C([C@H]([C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O)O...     89753   
2  C[C@]12CC[C@H]3[C@H]([C@@H]1CC[C@@H]2OC(=O)CCC...      9403   
3         C[C@@]12CC[C@@H](C1(C)C)C[C@H]2OC(=O)CSC#N  13218779   
4                               CC(=CCC/C(=C/CO)/C)C    637566   

                                        clean_smiles  
0         O=C(CO)[C@@H](O)[C@H](O)[C@@H](O)C(=O)[O-]  
1       O=C([O-])[C@H](O)[C@@H](O)[C@H](O)[C@H](O)CO  
2  C[C@]12CC[C@@H]3c4ccc(O)cc4CC[C@H]3[C@@H]1CC[C...  
3       CC1(C)[C@@H]2CC[C@@]1(C)[C@H](OC(=O)CSC#N)C2  
4                                CC(C)=CCC/C(C)=C/CO  
(6793, 3)
</pre></div>
</div>
</div>
</div>
<div class="alert alert-block alert-info">
<strong>Check Your Understanding</strong>
<p>
In this section, we transformed the original SMILES strings into <span>clean_smiles</span> by stripping salts and retaining only the largest fragment.
This step is often treated as “routine preprocessing,” but it is better understood as a deliberate modeling decision that changes the data we carry forward into fingerprint generation. The goal here is not to memorize a single correct approach, but to recognize that multiple reasonable approaches exist, and that the approach you choose should be justified and recorded. You also need to keep in mind that this decision does not only affect the data we are using now, but that any future data set using the model we build using these fingerprints will implicitly assume that data is processed in the same way.
</p>
<div style="
  background-color: #efffff;
  color: #000000;
  padding: 10px;
  border-radius: 4px;
  border: 1px solid #dddddd;
  margin-top: 10px;
">
<details>
<p><strong>What assumptions did we make? — recognizing that a decision was made</strong></p>
<ul>
  <li>What is more important for modeling purposes: the full formulation, or the parent structure?</li>
  <li>What assumptions are embedded in the phrase “biologically relevant parent compound”?</li>
  <li>How might the meaning of “clean” change depending on the assay context, the intended model, or the chemistry?</li>
</ul>
<p><strong>Why documentation may matter more than the decision itself</strong></p>
<ul>
  <li>If someone else reproduces this work, what do they need to know to obtain the same <span>clean_smiles</span> values?</li>
  <li>What information about the original records is lost once we overwrite or replace the original SMILES with cleaned structures?</li>
  <li>How could a future comparison between two models be confounded if they used different salt or mixture handling rules?</li>
  <li>Why is it important to keep versioned files as the dataset evolves through these transformations?</li>
</ul>
<p><strong>Looking ahead: applying the model to new data</strong></p>
<ul>
  <li>When this model is applied to new compounds, will those SMILES need to be cleaned in the same way?</li>
  <li>What could happen if new data explain salts or mixtures differently than the training data?</li>
  <li>How might inconsistent preprocessing affect fingerprint generation and model predictions?</li>
  <li>What parts of this preprocessing step should eventually be treated as part of the model pipeline rather than as one-time data cleaning?</li>
</ul>
<p>
The main takeaway is that preprocessing steps create a new version of your dataset and define the assumptions your model will carry forward.
Once fingerprints are generated, the model’s behavior depends not only on the algorithm, but also on whether future data are transformed in a compatible way.
</p>
</details>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_smiles</span><span class="p">[</span><span class="s1">&#39;is_identical&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_smiles</span><span class="p">[</span><span class="s1">&#39;smiles&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">df_smiles</span><span class="p">[</span><span class="s1">&#39;clean_smiles&#39;</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df_smiles</span><span class="p">[</span><span class="s1">&#39;is_identical&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span>
<span class="n">df_smiles</span><span class="p">[</span><span class="s1">&#39;is_identical&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>778
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>is_identical
False    6015
True      778
Name: count, dtype: int64
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># make sure the number of rows in df_smiles is equal to the number of unique CIDs we had previously</span>
<span class="n">df_activity</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;data/AID743139/curated/AID743139_Activity_CID_20260118_v1.csv&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Number of CIDs in df_smiles:&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">df_smiles</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Number of unique CIDs in original df:&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">df_activity</span><span class="p">[</span><span class="s1">&#39;cid&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()))</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The number of CIDs in df_smiles is equal to the number of unique CIDs in the original df:&quot;</span><span class="p">,</span>
       <span class="nb">len</span><span class="p">(</span><span class="n">df_smiles</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">df_activity</span><span class="p">[</span><span class="s1">&#39;cid&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()))</span>
<span class="n">df_smiles</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of CIDs in df_smiles: 6793
Number of unique CIDs in original df: 6793
The number of CIDs in df_smiles is equal to the number of unique CIDs in the original df: True
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>smiles</th>
      <th>cid</th>
      <th>clean_smiles</th>
      <th>is_identical</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>C(C(=O)[C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O.C(...</td>
      <td>12850184</td>
      <td>O=C(CO)[C@@H](O)[C@H](O)[C@@H](O)C(=O)[O-]</td>
      <td>False</td>
    </tr>
    <tr>
      <th>1</th>
      <td>C([C@H]([C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O)O...</td>
      <td>89753</td>
      <td>O=C([O-])[C@H](O)[C@@H](O)[C@H](O)[C@H](O)CO</td>
      <td>False</td>
    </tr>
    <tr>
      <th>2</th>
      <td>C[C@]12CC[C@H]3[C@H]([C@@H]1CC[C@@H]2OC(=O)CCC...</td>
      <td>9403</td>
      <td>C[C@]12CC[C@@H]3c4ccc(O)cc4CC[C@H]3[C@@H]1CC[C...</td>
      <td>False</td>
    </tr>
    <tr>
      <th>3</th>
      <td>C[C@@]12CC[C@@H](C1(C)C)C[C@H]2OC(=O)CSC#N</td>
      <td>13218779</td>
      <td>CC1(C)[C@@H]2CC[C@@]1(C)[C@H](OC(=O)CSC#N)C2</td>
      <td>False</td>
    </tr>
    <tr>
      <th>4</th>
      <td>CC(=CCC/C(=C/CO)/C)C</td>
      <td>637566</td>
      <td>CC(C)=CCC/C(C)=C/CO</td>
      <td>False</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="alert alert-block alert-info">
<strong>Check Your Understanding</strong>
<p>
After generating <span>clean_smiles</span>, we compared the original SMILES strings to the cleaned ones and found that only 778 entries were identical, while 6,015 were different.
At first glance, it is tempting to conclude that this means 6,015 compounds were “changed” during cleaning.
This check is useful, but it is also an opportunity to think carefully about what our comparison is actually measuring.
</p>
<div style="
  background-color: #efffff;
  color: #000000;
  padding: 10px;
  border-radius: 4px;
  border: 1px solid #dddddd;
  margin-top: 10px;
">
<details>
<p><strong>What question does this comparison really answer?</strong></p>
<ul>
  <li>Does a difference between <span>smiles</span> and <span>clean_smiles</span> always mean the underlying chemical structure changed?</li>
  <li>What assumptions are we making when we treat SMILES strings as if they were canonical identifiers?</li>
  <li>How might RDKit-generated SMILES differ from PubChem SMILES even when no salt or fragment was removed?</li>
</ul>
<p><strong>Why this result can be misleading at first glance</strong></p>
<ul>
  <li>SMILES are not unique representations; the same molecule can be written in multiple valid ways.</li>
  <li>RDKit often reorders atoms or normalizes bonding when generating SMILES, even if the structure is unchanged.</li>
  <li>As a result, a <span>False</span> value here may reflect a change in representation, not a change in chemistry.</li>
</ul>
<p><strong>Connecting this to data provenance and decision making</strong></p>
<ul>
  <li>What evidence do we actually need to support the claim that a compound was chemically altered during cleaning?</li>
  <li>How does the choice of representation affect our interpretation of “before” and “after” datasets?</li>
  <li>Why is it important to distinguish between a structural change and a formatting or normalization change?</li>
</ul>
<p><strong>Looking ahead to better comparisons</strong></p>
<ul>
  <li>What properties of a molecule should remain invariant if the underlying structure is truly the same?</li>
  <li>How might a representation designed to be canonical help answer this question more reliably?</li>
  <li>In later steps, how could misleading checks like this propagate confusion if they are not carefully interpreted?</li>
</ul>
<p>
The key takeaway is that comparing SMILES strings is a convenient first check, but it is not a definitive test of whether a compound changed during preprocessing.
Understanding what a comparison actually measures is just as important as running the comparison itself, especially when those results inform downstream modeling decisions.
</p>
</details>
</div>
</div>
</section>
<section id="comparing-structures-using-inchi">
<h2>5.2 Comparing structures using InChI<a class="headerlink" href="#comparing-structures-using-inchi" title="Link to this heading">#</a></h2>
<p>In the following code cell we will convert both the original and cleaned SMILES strings to InChIKeys using RDKit. InChIKeys are unique, fixed-length identifiers derived from the International Chemical Identifier (InChI) representation of a molecule. They provide a standardized way to compare chemical structures, as two identical molecules will produce the same InChIKey.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">rdkit</span><span class="w"> </span><span class="kn">import</span> <span class="n">Chem</span>

<span class="k">def</span><span class="w"> </span><span class="nf">smiles_to_inchi</span><span class="p">(</span><span class="n">sm</span><span class="p">):</span>
    <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromSmiles</span><span class="p">(</span><span class="n">sm</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">mol</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="k">return</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolToInchi</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>

<span class="k">def</span><span class="w"> </span><span class="nf">smiles_to_inchikey</span><span class="p">(</span><span class="n">sm</span><span class="p">):</span>
    <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromSmiles</span><span class="p">(</span><span class="n">sm</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">mol</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="k">return</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolToInchiKey</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>

<span class="n">df_smiles</span><span class="p">[</span><span class="s1">&#39;inchikey_original&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_smiles</span><span class="p">[</span><span class="s1">&#39;smiles&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">smiles_to_inchikey</span><span class="p">)</span>
<span class="n">df_smiles</span><span class="p">[</span><span class="s1">&#39;inchikey_clean&#39;</span><span class="p">]</span>    <span class="o">=</span> <span class="n">df_smiles</span><span class="p">[</span><span class="s1">&#39;clean_smiles&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">smiles_to_inchikey</span><span class="p">)</span>

<span class="n">df_smiles</span><span class="p">[</span><span class="s1">&#39;is_identical&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">df_smiles</span><span class="p">[</span><span class="s1">&#39;inchikey_original&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">df_smiles</span><span class="p">[</span><span class="s1">&#39;inchikey_clean&#39;</span><span class="p">]</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">df_smiles</span><span class="p">[</span><span class="s1">&#39;is_identical&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span>
<span class="n">df_smiles</span><span class="p">[</span><span class="s1">&#39;is_identical&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[08:54:53] WARNING: not removing hydrogen atom without neighbors
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>5336
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>is_identical
True     5336
False    1457
Name: count, dtype: int64
</pre></div>
</div>
</div>
</div>
<div class="alert alert-block alert-info">
<strong>Check Your Understanding</strong>
<p>
We repeated the comparison using InChI-derived identifiers and found that 5,336 entries were identical, while 1,457 were different. Discuss why these result looks very different from the SMILES-based comparison and points to an important concept in cheminformatics: not all molecular representations are intended to answer the same question.
</p>
<div style="
  background-color: #efffff;
  color: #000000;
  padding: 10px;
  border-radius: 4px;
  border: 1px solid #dddddd;
  margin-top: 10px;
">
<details>
<p><strong>Why does the InChI comparison behave differently?</strong></p>
<ul>
  <li>InChI (International Chemical Identifier) is an IUPAC-led, open-source standard designed to uniquely describe a chemical structure.</li>
  <li>Two molecules that are chemically identical should, by design, generate the same InChI, even if their SMILES strings differ.</li>
  <li>This makes InChI (or InChIKey) much better suited for identity checks than raw SMILES strings.</li>
</ul>
<p><strong>What this tells us about SMILES</strong></p>
<ul>
  <li>SMILES are flexible, human-readable encodings, not unique identifiers.</li>
  <li>There is no single, universal “canonical SMILES” standard; different software packages use different canonicalization algorithms.</li>
  <li>As a result, two SMILES strings can differ even when they describe the same molecule.</li>
</ul>
<p>
A well-known illustration of this flexibility is that there are thousands of valid SMILES representations for a single molecule such as caffeine.
For a fun and eye-opening example, see this blog post:
<a href="https://nextmovesoftware.com/blog/2014/07/15/how-do-i-write-thee-let-me-count-the-ways/" target="_blank">
How do I write thee? Let me count the ways
</a>.
</p>
<p><strong>Key takeaway</strong></p>
<ul>
  <li>SMILES are excellent for storing structures and generating fingerprints, but they are not reliable for determining molecular identity.</li>
  <li>InChI was created specifically to solve this problem and should be used when the question is “are these two molecules the same?”</li>
  <li>Choosing the right representation is not a modeling decision that changes the data, but a correctness issue that affects how results are interpreted.</li>
</ul>
<p>
The main lesson here is that tools in cheminformatics are purpose-built.
Understanding what question a representation is designed to answer is essential for using it correctly and avoiding misleading conclusions.
</p>
</details>
</div>
</div>
<p>Our last step before fingerprint generation will be to remove the series we generated to analyze the data and save the CID and clean_smiles for creating fingerprints with.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_smiles</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>smiles</th>
      <th>cid</th>
      <th>clean_smiles</th>
      <th>is_identical</th>
      <th>inchikey_original</th>
      <th>inchikey_clean</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>C(C(=O)[C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O.C(...</td>
      <td>12850184</td>
      <td>O=C(CO)[C@@H](O)[C@H](O)[C@@H](O)C(=O)[O-]</td>
      <td>False</td>
      <td>HNXKHQGRMMZMKP-UINDWOIOSA-L</td>
      <td>IZSRJDGCGRAUAR-MROZADKFSA-M</td>
    </tr>
    <tr>
      <th>1</th>
      <td>C([C@H]([C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O)O...</td>
      <td>89753</td>
      <td>O=C([O-])[C@H](O)[C@@H](O)[C@H](O)[C@H](O)CO</td>
      <td>False</td>
      <td>PECPTSUHUFWKDG-IYEMJOQQSA-L</td>
      <td>RGHNJXZEOKUKBD-SQOUGZDYSA-M</td>
    </tr>
    <tr>
      <th>2</th>
      <td>C[C@]12CC[C@H]3[C@H]([C@@H]1CC[C@@H]2OC(=O)CCC...</td>
      <td>9403</td>
      <td>C[C@]12CC[C@@H]3c4ccc(O)cc4CC[C@H]3[C@@H]1CC[C...</td>
      <td>True</td>
      <td>UOACKFBJUYNSLK-XRKIENNPSA-N</td>
      <td>UOACKFBJUYNSLK-XRKIENNPSA-N</td>
    </tr>
    <tr>
      <th>3</th>
      <td>C[C@@]12CC[C@@H](C1(C)C)C[C@H]2OC(=O)CSC#N</td>
      <td>13218779</td>
      <td>CC1(C)[C@@H]2CC[C@@]1(C)[C@H](OC(=O)CSC#N)C2</td>
      <td>True</td>
      <td>IXEVGHXRXDBAOB-BREBYQMCSA-N</td>
      <td>IXEVGHXRXDBAOB-BREBYQMCSA-N</td>
    </tr>
    <tr>
      <th>4</th>
      <td>CC(=CCC/C(=C/CO)/C)C</td>
      <td>637566</td>
      <td>CC(C)=CCC/C(C)=C/CO</td>
      <td>True</td>
      <td>GLZPCOQZEFWAFX-JXMROGBWSA-N</td>
      <td>GLZPCOQZEFWAFX-JXMROGBWSA-N</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># reorder columns to have &#39;cid&#39; first, then &#39;clean_smiles&#39; and remove columns used to check for differences</span>

<span class="n">df_smiles</span> <span class="o">=</span> <span class="n">df_smiles</span><span class="p">[[</span><span class="s1">&#39;cid&#39;</span><span class="p">,</span><span class="s1">&#39;clean_smiles&#39;</span><span class="p">]]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df_smiles</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df_smiles</span><span class="o">.</span><span class="n">dtypes</span><span class="p">)</span>
<span class="n">df_smiles</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(6793, 2)
cid              int64
clean_smiles    object
dtype: object
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cid</th>
      <th>clean_smiles</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>12850184</td>
      <td>O=C(CO)[C@@H](O)[C@H](O)[C@@H](O)C(=O)[O-]</td>
    </tr>
    <tr>
      <th>1</th>
      <td>89753</td>
      <td>O=C([O-])[C@H](O)[C@@H](O)[C@H](O)[C@H](O)CO</td>
    </tr>
    <tr>
      <th>2</th>
      <td>9403</td>
      <td>C[C@]12CC[C@@H]3c4ccc(O)cc4CC[C@H]3[C@@H]1CC[C...</td>
    </tr>
    <tr>
      <th>3</th>
      <td>13218779</td>
      <td>CC1(C)[C@@H]2CC[C@@]1(C)[C@H](OC(=O)CSC#N)C2</td>
    </tr>
    <tr>
      <th>4</th>
      <td>637566</td>
      <td>CC(C)=CCC/C(C)=C/CO</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<section id="save-as-aid743139-nosalt-pcsmiles">
<h3>Save as AID743139_noSalt_pcSMILES…<a class="headerlink" href="#save-as-aid743139-nosalt-pcsmiles" title="Link to this heading">#</a></h3>
<p>Save the dataframe in your curated data folder with date stamp and version number using <code class="docutils literal notranslate"><span class="pre">prefix=&quot;AID743139_noSalt_pcSMILES&quot;,</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># write your code here:</span>
<span class="c1"># Save to data/ with versioning</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">cinf26pk.core</span><span class="w"> </span><span class="kn">import</span> <span class="n">filenames</span>
<span class="c1"># generate path</span>
<span class="n">PROJECT_ROOT</span> <span class="o">=</span> <span class="n">Path</span><span class="o">.</span><span class="n">cwd</span><span class="p">()</span> <span class="c1"># project root (10_SupervisedML)</span>
<span class="n">CUR</span> <span class="o">=</span> <span class="n">PROJECT_ROOT</span> <span class="o">/</span> <span class="s2">&quot;data&quot;</span> <span class="o">/</span> <span class="s2">&quot;AID743139&quot;</span> <span class="o">/</span> <span class="s2">&quot;curated&quot;</span>
<span class="n">CUR</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Create deterministic filename with date stamp</span>
<span class="n">fname</span> <span class="o">=</span> <span class="n">filenames</span><span class="o">.</span><span class="n">make_filename</span><span class="p">(</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;AID743139_noSalt_pcSMILES&quot;</span><span class="p">,</span>
    <span class="n">ext</span><span class="o">=</span><span class="s2">&quot;csv&quot;</span><span class="p">,</span>
    <span class="n">folder</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">CUR</span><span class="p">)</span>
<span class="p">)</span>

<span class="c1"># Save raw dataset</span>
<span class="n">df_smiles</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">CUR</span><span class="o">/</span><span class="n">fname</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[Saved] Curated PubChem data → </span><span class="si">{</span><span class="n">fname</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[Saved] Curated PubChem data → /home/rebelford/jupyterbooks/cinf26book/content/modules/10_SupervisedML/data/AID743139/curated/AID743139_noSalt_pcSMILES_20260118_v1.csv
</pre></div>
</div>
</div>
</div>
<div class="alert alert-block alert-info">
<strong>Check your understanding</strong>
<ol class="arabic simple">
<li><p>When we requested the SMILES data from PubChem through PUG REST, why did we chunk our data and use time.sleep(.2)?</p></li>
<li><p>Briefly explain the steps and importance for cleaning the data in the notebook thus far.</p></li>
<li><p>How many compounds were salts or multicomponent mixtures and how did we handle them?</p></li>
</ol>
  <div style="
    background-color: #efffff;
    color: #000000;
    padding: 10px;
    border-radius: 4px;
    border: 1px solid #dddddd;
    margin-top: 10px;
  ">
<details>
    <summary>Answer</summary>
<ol class="arabic simple">
<li><p>As noted in the Introduction to PUG REST notebook, PubChem enforces 30-second time limit per request. If we attempt fetch properties for too many compounds at once, the request may exceed this limit and result in a timeout error. Since our dataset includes nearly 7000 molecules, we will split the list of CIDs into smaller chunks and process each chunk separately. This approach helps keep each request efficient, reduces the likelihood of failure, and ensures smooth and successful data retrieval from the API.</p></li>
<li><p>Cleaning the data has involved several key steps: removing entries without CIDs (since structural information cannot be retrieved without them), resolving conflicting activity labels to ensure each compound is consistently classified, and converting text-based activity labels into numerical values suitable for machine learning. These steps are critical to ensure that the model is trained on accurate, consistent, and well-structured data, which improves its performance and prevents errors during modeling.</p></li>
<li><p>When we compared Smiles from PubChem and those after salt removal that were generated by RDKit we found 6013 that were not equal.  But Smiles are not canonical and RDKit and PubChem could generate different SMILES for the same molecule.  So we then converted them to InChIKeys, which are canonical, and found that 1,457 had been changed.  For these compounds we extracted the fragment with the largest number of heavy atoms, assuming this represented the parent compound.</p></li>
</ol>
</details>
</div>
</div>
</section>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="convert-smiles-to-binary-data-for-model-input">
<h1>6 Convert SMILES to Binary Data for Model Input<a class="headerlink" href="#convert-smiles-to-binary-data-for-model-input" title="Link to this heading">#</a></h1>
<section id="generate-maccs-keys-from-smiles">
<h2>6.1 Generate MACCS keys from SMILES<a class="headerlink" href="#generate-maccs-keys-from-smiles" title="Link to this heading">#</a></h2>
<p>At this point in the workflow, we have completed the chemical and biological curation of the dataset.
We have:</p>
<ul class="simple">
<li><p>Selected which compounds to include</p></li>
<li><p>Resolved conflicting assay outcomes</p></li>
<li><p>Defined a binary activity label</p></li>
<li><p>Standardized molecular representations by removing salts and mixtures</p></li>
</ul>
<p>The remaining task in this module is to translate chemical structure into a numerical form that a machine-learning algorithm can consume. To do this, we convert each cleaned SMILES string into an RDKit molecule object and generate MACCS structural keys. Each MACCS key encodes the presence (<code class="docutils literal notranslate"><span class="pre">1</span></code>) or absence (<code class="docutils literal notranslate"><span class="pre">0</span></code>) of a predefined substructure pattern. The result is a fixed-length binary vector that represents chemical structure in a form suitable for supervised learning. Rather than working with RDKit bitvector objects directly, we explicitly convert these fingerprints into tabular numerical data and save them to disk. The resulting file contains:</p>
<ul class="simple">
<li><p>Compound identifiers (<code class="docutils literal notranslate"><span class="pre">cid</span></code>)</p></li>
<li><p>The curated binary activity label (<code class="docutils literal notranslate"><span class="pre">activity</span></code>)</p></li>
<li><p>The cleaned SMILES representation (<code class="docutils literal notranslate"><span class="pre">clean_smiles</span></code>)</p></li>
<li><p>One column for each MACCS key (e.g., <code class="docutils literal notranslate"><span class="pre">MACCS000</span></code> through <code class="docutils literal notranslate"><span class="pre">MACCS166</span></code>)</p></li>
</ul>
<p>This file represents the <strong>final output of Module 10.1</strong>.
It serves as the starting point for:</p>
<ul class="simple">
<li><p>Model construction in <em>10.2 Naive Bayes: Model Construction Workflow</em></p></li>
<li><p>Pipeline construction, reuse, and inference in <em>10.3 Pipelines and Inference</em></p></li>
</ul>
<p>Once fingerprints are generated, the model no longer “sees” molecules, it only sees numbers. Any chemical assumptions made earlier (salt removal, fragment selection, fingerprint choice) are now frozen into the feature representation. This is why the outputs of this step are treated as durable artifacts and reused across later modeling workflows. By saving this representation explicitly, we make the transition from “chemical data preparation” to “machine-learning experimentation” clear, reproducible, and inspectable. Our directory architecture will support this by moving the fingerprints out of the <code class="docutils literal notranslate"><span class="pre">\data\curated</span></code> and into a new <code class="docutils literal notranslate"><span class="pre">\data\features</span></code> directory.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># load ..._noSalt_pcSMILES... into df_smiles if you restarted the kernel</span>
<span class="n">df_smiles</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cid</th>
      <th>clean_smiles</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>12850184</td>
      <td>O=C(CO)[C@@H](O)[C@H](O)[C@@H](O)C(=O)[O-]</td>
    </tr>
    <tr>
      <th>1</th>
      <td>89753</td>
      <td>O=C([O-])[C@H](O)[C@@H](O)[C@H](O)[C@H](O)CO</td>
    </tr>
    <tr>
      <th>2</th>
      <td>9403</td>
      <td>C[C@]12CC[C@@H]3c4ccc(O)cc4CC[C@H]3[C@@H]1CC[C...</td>
    </tr>
    <tr>
      <th>3</th>
      <td>13218779</td>
      <td>CC1(C)[C@@H]2CC[C@@]1(C)[C@H](OC(=O)CSC#N)C2</td>
    </tr>
    <tr>
      <th>4</th>
      <td>637566</td>
      <td>CC(C)=CCC/C(C)=C/CO</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">rdkit</span><span class="w"> </span><span class="kn">import</span> <span class="n">Chem</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">rdkit.Chem</span><span class="w"> </span><span class="kn">import</span> <span class="n">MACCSkeys</span>

<span class="n">fps</span><span class="o">=</span><span class="nb">dict</span><span class="p">()</span>

<span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df_smiles</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()</span> <span class="p">:</span>
    
    <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolFromSmiles</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">clean_smiles</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="n">mol</span> <span class="o">==</span> <span class="kc">None</span> <span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Can&#39;t generate MOL object:&quot;</span><span class="p">,</span> <span class="s2">&quot;CID&quot;</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">cid</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">smiles</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">fps</span><span class="p">[</span><span class="n">row</span><span class="o">.</span><span class="n">cid</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">row</span><span class="o">.</span><span class="n">cid</span><span class="p">]</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">MACCSkeys</span><span class="o">.</span><span class="n">GenMACCSKeys</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span><span class="o">.</span><span class="n">ToBitString</span><span class="p">())</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Dictionary contains </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">fps</span><span class="p">)</span><span class="si">}</span><span class="s2"> key–value pairs&quot;</span><span class="p">)</span>

<span class="c1"># Print first two key-value pairs of the dictionary fps</span>
<span class="k">def</span><span class="w"> </span><span class="nf">dict_head</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">2</span><span class="p">):</span>
    <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">items</span><span class="p">())[:</span><span class="n">n</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">dict_head</span><span class="p">(</span><span class="n">fps</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Dictionary contains 6793 key–value pairs
[(12850184, [12850184, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;1&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;1&#39;, &#39;1&#39;, &#39;1&#39;, &#39;0&#39;, &#39;1&#39;, &#39;0&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;]), (89753, [89753, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;1&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;1&#39;, &#39;1&#39;, &#39;1&#39;, &#39;0&#39;, &#39;1&#39;, &#39;0&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;1&#39;, &#39;0&#39;, &#39;0&#39;])]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a dataframe of the fingerprints generated fromm the noSalt smiles</span>
<span class="c1"># Note we lost the fingerprints</span>
<span class="c1"># Generate column names</span>
<span class="n">fpbitnames</span> <span class="o">=</span> <span class="p">[]</span>

<span class="n">fpbitnames</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;cid&#39;</span><span class="p">)</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">167</span><span class="p">):</span>   <span class="c1"># from MACCS000 to MACCS166</span>
    <span class="n">fpbitnames</span><span class="o">.</span><span class="n">append</span><span class="p">(</span> <span class="s2">&quot;maccs&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="o">.</span><span class="n">zfill</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span> <span class="p">)</span>

<span class="n">df_fps</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">fps</span><span class="p">,</span> <span class="n">orient</span><span class="o">=</span><span class="s1">&#39;index&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">fpbitnames</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">df_fps</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="n">df_fps</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(6793, 168)
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cid</th>
      <th>maccs000</th>
      <th>maccs001</th>
      <th>maccs002</th>
      <th>maccs003</th>
      <th>maccs004</th>
      <th>maccs005</th>
      <th>maccs006</th>
      <th>maccs007</th>
      <th>maccs008</th>
      <th>...</th>
      <th>maccs157</th>
      <th>maccs158</th>
      <th>maccs159</th>
      <th>maccs160</th>
      <th>maccs161</th>
      <th>maccs162</th>
      <th>maccs163</th>
      <th>maccs164</th>
      <th>maccs165</th>
      <th>maccs166</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>12850184</th>
      <td>12850184</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>89753</th>
      <td>89753</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>9403</th>
      <td>9403</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
    </tr>
    <tr>
      <th>13218779</th>
      <td>13218779</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
    </tr>
    <tr>
      <th>637566</th>
      <td>637566</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 168 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Merge DataFrames (add fingerprints to end of df)</span>
<span class="n">df_fps</span> <span class="o">=</span> <span class="n">df_fps</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
    <span class="n">df_smiles</span><span class="p">[[</span><span class="s1">&#39;cid&#39;</span><span class="p">,</span> <span class="s1">&#39;clean_smiles&#39;</span><span class="p">]],</span>
    <span class="n">on</span><span class="o">=</span><span class="s1">&#39;cid&#39;</span><span class="p">,</span>
    <span class="n">how</span><span class="o">=</span><span class="s1">&#39;left&#39;</span>
<span class="p">)</span>
<span class="n">df_fps</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cid</th>
      <th>maccs000</th>
      <th>maccs001</th>
      <th>maccs002</th>
      <th>maccs003</th>
      <th>maccs004</th>
      <th>maccs005</th>
      <th>maccs006</th>
      <th>maccs007</th>
      <th>maccs008</th>
      <th>...</th>
      <th>maccs158</th>
      <th>maccs159</th>
      <th>maccs160</th>
      <th>maccs161</th>
      <th>maccs162</th>
      <th>maccs163</th>
      <th>maccs164</th>
      <th>maccs165</th>
      <th>maccs166</th>
      <th>clean_smiles</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>12850184</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>O=C(CO)[C@@H](O)[C@H](O)[C@@H](O)C(=O)[O-]</td>
    </tr>
    <tr>
      <th>1</th>
      <td>89753</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>O=C([O-])[C@H](O)[C@@H](O)[C@H](O)[C@H](O)CO</td>
    </tr>
    <tr>
      <th>2</th>
      <td>9403</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
      <td>C[C@]12CC[C@@H]3c4ccc(O)cc4CC[C@H]3[C@@H]1CC[C...</td>
    </tr>
  </tbody>
</table>
<p>3 rows × 169 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Arrange Columns (move fingerprints to left)</span>
<span class="n">front</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;cid&#39;</span><span class="p">,</span> <span class="s1">&#39;clean_smiles&#39;</span><span class="p">]</span>
<span class="n">rest</span> <span class="o">=</span> <span class="p">[</span><span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">df_fps</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">c</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">front</span><span class="p">]</span>

<span class="n">df_fps</span> <span class="o">=</span> <span class="n">df_fps</span><span class="p">[</span><span class="n">front</span> <span class="o">+</span> <span class="n">rest</span><span class="p">]</span>

<span class="n">df_fps</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cid</th>
      <th>clean_smiles</th>
      <th>maccs000</th>
      <th>maccs001</th>
      <th>maccs002</th>
      <th>maccs003</th>
      <th>maccs004</th>
      <th>maccs005</th>
      <th>maccs006</th>
      <th>maccs007</th>
      <th>...</th>
      <th>maccs157</th>
      <th>maccs158</th>
      <th>maccs159</th>
      <th>maccs160</th>
      <th>maccs161</th>
      <th>maccs162</th>
      <th>maccs163</th>
      <th>maccs164</th>
      <th>maccs165</th>
      <th>maccs166</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>12850184</td>
      <td>O=C(CO)[C@@H](O)[C@H](O)[C@@H](O)C(=O)[O-]</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>89753</td>
      <td>O=C([O-])[C@H](O)[C@@H](O)[C@H](O)[C@H](O)CO</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>9403</td>
      <td>C[C@]12CC[C@@H]3c4ccc(O)cc4CC[C@H]3[C@@H]1CC[C...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
<p>3 rows × 169 columns</p>
</div></div></div>
</div>
<section id="save-as-features-aid743139-maccs-nosalt">
<h3>Save as features/AID743139_MACCS_noSalt…<a class="headerlink" href="#save-as-features-aid743139-maccs-nosalt" title="Link to this heading">#</a></h3>
<p>Lets now save this in a new folder, data/features and give it a versioning name AID743139_MACCS_noSalt</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Save to data/ with versioning</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">cinf26pk.core</span><span class="w"> </span><span class="kn">import</span> <span class="n">filenames</span>
<span class="c1"># generate path</span>
<span class="n">PROJECT_ROOT</span> <span class="o">=</span> <span class="n">Path</span><span class="o">.</span><span class="n">cwd</span><span class="p">()</span> <span class="c1"># project root (10_SupervisedML)</span>
<span class="n">CUR</span> <span class="o">=</span> <span class="n">PROJECT_ROOT</span> <span class="o">/</span> <span class="s2">&quot;data&quot;</span> <span class="o">/</span> <span class="s2">&quot;AID743139&quot;</span> <span class="o">/</span> <span class="s2">&quot;features&quot;</span>
<span class="n">CUR</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Create deterministic filename with date stamp</span>
<span class="n">fname</span> <span class="o">=</span> <span class="n">filenames</span><span class="o">.</span><span class="n">make_filename</span><span class="p">(</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;AID743139_MACCS_noSalt&quot;</span><span class="p">,</span>
    <span class="n">ext</span><span class="o">=</span><span class="s2">&quot;csv&quot;</span><span class="p">,</span>
    <span class="n">folder</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">CUR</span><span class="p">)</span>
<span class="p">)</span>

<span class="c1"># Save raw dataset</span>
<span class="n">df_fps</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">CUR</span><span class="o">/</span><span class="n">fname</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[Saved] Curated PubChem data → </span><span class="si">{</span><span class="n">fname</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[Saved] Curated PubChem data → /home/rebelford/jupyterbooks/cinf26book/content/modules/10_SupervisedML/data/AID743139/features/AID743139_MACCS_noSalt_20260118_v1.csv
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="merge-activity-data-and-fingerprint-information">
<h2>6.2 Merge activity data and fingerprint information<a class="headerlink" href="#merge-activity-data-and-fingerprint-information" title="Link to this heading">#</a></h2>
<p>note, you may need to load the data into df_activity, which you generated in step 4.6.1 (with a file prefix of AID743139_Activity_CID_)</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_activity</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;data/AID743139/curated/AID743139_Activity_CID_20260118_v1.csv&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">df_activity</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="n">df_activity</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(6793, 3)
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cid</th>
      <th>activity</th>
      <th>PUBCHEM_EXT_DATASOURCE_SMILES</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>12850184</td>
      <td>0</td>
      <td>C(C(=O)[C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O.C(...</td>
    </tr>
    <tr>
      <th>1</th>
      <td>89753</td>
      <td>0</td>
      <td>C([C@H]([C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O)O...</td>
    </tr>
    <tr>
      <th>2</th>
      <td>9403</td>
      <td>0</td>
      <td>C[C@]12CC[C@H]3[C@H]([C@@H]1CC[C@@H]2OC(=O)CCC...</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">df_fps</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="n">df_fps</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(6793, 169)
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cid</th>
      <th>clean_smiles</th>
      <th>maccs000</th>
      <th>maccs001</th>
      <th>maccs002</th>
      <th>maccs003</th>
      <th>maccs004</th>
      <th>maccs005</th>
      <th>maccs006</th>
      <th>maccs007</th>
      <th>...</th>
      <th>maccs157</th>
      <th>maccs158</th>
      <th>maccs159</th>
      <th>maccs160</th>
      <th>maccs161</th>
      <th>maccs162</th>
      <th>maccs163</th>
      <th>maccs164</th>
      <th>maccs165</th>
      <th>maccs166</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>12850184</td>
      <td>O=C(CO)[C@@H](O)[C@H](O)[C@@H](O)C(=O)[O-]</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>89753</td>
      <td>O=C([O-])[C@H](O)[C@@H](O)[C@H](O)[C@H](O)CO</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>9403</td>
      <td>C[C@]12CC[C@@H]3c4ccc(O)cc4CC[C@H]3[C@@H]1CC[C...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
<p>3 rows × 169 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># For each row in df_activity, </span>
<span class="c1">#attach the fingerprint columns from df_fps that have the same CID</span>
<span class="n">df_data</span> <span class="o">=</span> <span class="n">df_activity</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">df_fps</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;cid&#39;</span><span class="p">),</span> <span class="n">on</span><span class="o">=</span><span class="s1">&#39;cid&#39;</span><span class="p">)</span>
<span class="n">df_data</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cid</th>
      <th>activity</th>
      <th>PUBCHEM_EXT_DATASOURCE_SMILES</th>
      <th>clean_smiles</th>
      <th>maccs000</th>
      <th>maccs001</th>
      <th>maccs002</th>
      <th>maccs003</th>
      <th>maccs004</th>
      <th>maccs005</th>
      <th>...</th>
      <th>maccs157</th>
      <th>maccs158</th>
      <th>maccs159</th>
      <th>maccs160</th>
      <th>maccs161</th>
      <th>maccs162</th>
      <th>maccs163</th>
      <th>maccs164</th>
      <th>maccs165</th>
      <th>maccs166</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>12850184</td>
      <td>0</td>
      <td>C(C(=O)[C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O.C(...</td>
      <td>O=C(CO)[C@@H](O)[C@H](O)[C@@H](O)C(=O)[O-]</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>89753</td>
      <td>0</td>
      <td>C([C@H]([C@H]([C@@H]([C@H](C(=O)[O-])O)O)O)O)O...</td>
      <td>O=C([O-])[C@H](O)[C@@H](O)[C@H](O)[C@H](O)CO</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>9403</td>
      <td>0</td>
      <td>C[C@]12CC[C@H]3[C@H]([C@@H]1CC[C@@H]2OC(=O)CCC...</td>
      <td>C[C@]12CC[C@@H]3c4ccc(O)cc4CC[C@H]3[C@@H]1CC[C...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
<p>3 rows × 171 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># drop pubchem source smiles</span>
<span class="n">df_data</span> <span class="o">=</span> <span class="n">df_data</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;PUBCHEM_EXT_DATASOURCE_SMILES&#39;</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df_data</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="n">df_data</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(6793, 170)
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cid</th>
      <th>activity</th>
      <th>clean_smiles</th>
      <th>maccs000</th>
      <th>maccs001</th>
      <th>maccs002</th>
      <th>maccs003</th>
      <th>maccs004</th>
      <th>maccs005</th>
      <th>maccs006</th>
      <th>...</th>
      <th>maccs157</th>
      <th>maccs158</th>
      <th>maccs159</th>
      <th>maccs160</th>
      <th>maccs161</th>
      <th>maccs162</th>
      <th>maccs163</th>
      <th>maccs164</th>
      <th>maccs165</th>
      <th>maccs166</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>12850184</td>
      <td>0</td>
      <td>O=C(CO)[C@@H](O)[C@H](O)[C@@H](O)C(=O)[O-]</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>89753</td>
      <td>0</td>
      <td>O=C([O-])[C@H](O)[C@@H](O)[C@H](O)[C@H](O)CO</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>9403</td>
      <td>0</td>
      <td>C[C@]12CC[C@@H]3c4ccc(O)cc4CC[C@H]3[C@@H]1CC[C...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
<p>3 rows × 170 columns</p>
</div></div></div>
</div>
<p>Check to see if there are any CIDs for which the MACCS keys could not be generated.  They need to be removed from <strong>df_data</strong>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_data</span><span class="p">[</span><span class="n">df_data</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cid</th>
      <th>activity</th>
      <th>clean_smiles</th>
      <th>maccs000</th>
      <th>maccs001</th>
      <th>maccs002</th>
      <th>maccs003</th>
      <th>maccs004</th>
      <th>maccs005</th>
      <th>maccs006</th>
      <th>...</th>
      <th>maccs157</th>
      <th>maccs158</th>
      <th>maccs159</th>
      <th>maccs160</th>
      <th>maccs161</th>
      <th>maccs162</th>
      <th>maccs163</th>
      <th>maccs164</th>
      <th>maccs165</th>
      <th>maccs166</th>
    </tr>
  </thead>
  <tbody>
  </tbody>
</table>
<p>0 rows × 170 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;before drop nan: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">df_data</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">df_data</span> <span class="o">=</span> <span class="n">df_data</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;after drop nan: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">df_data</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">df_data</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>before drop nan: 6793
after drop nan: 6793
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cid</th>
      <th>activity</th>
      <th>clean_smiles</th>
      <th>maccs000</th>
      <th>maccs001</th>
      <th>maccs002</th>
      <th>maccs003</th>
      <th>maccs004</th>
      <th>maccs005</th>
      <th>maccs006</th>
      <th>...</th>
      <th>maccs157</th>
      <th>maccs158</th>
      <th>maccs159</th>
      <th>maccs160</th>
      <th>maccs161</th>
      <th>maccs162</th>
      <th>maccs163</th>
      <th>maccs164</th>
      <th>maccs165</th>
      <th>maccs166</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>12850184</td>
      <td>0</td>
      <td>O=C(CO)[C@@H](O)[C@H](O)[C@@H](O)C(=O)[O-]</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>89753</td>
      <td>0</td>
      <td>O=C([O-])[C@H](O)[C@@H](O)[C@H](O)[C@H](O)CO</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>9403</td>
      <td>0</td>
      <td>C[C@]12CC[C@@H]3c4ccc(O)cc4CC[C@H]3[C@@H]1CC[C...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
<p>3 rows × 170 columns</p>
</div></div></div>
</div>
<p>Save df_data in CSV for future use using the prefix=”AID743139_MACCS_activites_noSalt”</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Save to data/ with versioning</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">cinf26pk.core</span><span class="w"> </span><span class="kn">import</span> <span class="n">filenames</span>
<span class="c1"># generate path</span>
<span class="n">PROJECT_ROOT</span> <span class="o">=</span> <span class="n">Path</span><span class="o">.</span><span class="n">cwd</span><span class="p">()</span> <span class="c1"># project root (10_SupervisedML)</span>
<span class="n">FEAT</span> <span class="o">=</span> <span class="n">PROJECT_ROOT</span> <span class="o">/</span> <span class="s2">&quot;data&quot;</span> <span class="o">/</span> <span class="s2">&quot;AID743139&quot;</span> <span class="o">/</span> <span class="s2">&quot;features&quot;</span>
<span class="n">FEAT</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Create deterministic filename with date stamp</span>
<span class="n">fname</span> <span class="o">=</span> <span class="n">filenames</span><span class="o">.</span><span class="n">make_filename</span><span class="p">(</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;AID743139_MACCS_activites_noSalt&quot;</span><span class="p">,</span>
    <span class="n">ext</span><span class="o">=</span><span class="s2">&quot;csv&quot;</span><span class="p">,</span>
    <span class="n">folder</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">FEAT</span><span class="p">)</span>
<span class="p">)</span>

<span class="c1"># Save raw dataset</span>
<span class="n">df_data</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">FEAT</span><span class="o">/</span><span class="n">fname</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[Saved] Features csv data → </span><span class="si">{</span><span class="n">fname</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[Saved] Features csv data → /home/rebelford/jupyterbooks/cinf26book/content/modules/10_SupervisedML/data/AID743139/features/AID743139_MACCS_activites_noSalt_20260118_v1.csv
</pre></div>
</div>
</div>
</div>
<p>At this point, we have completed the primary objective of this module: transforming curated chemical and biological data into a feature-level dataset suitable for supervised machine learning. The file saved in <code class="docutils literal notranslate"><span class="pre">data/AID743139/features/</span></code> captures the outcome of all decisions made in this module—compound selection, activity labeling, structural standardization, and fingerprint representation—encoded as numerical input that models can consume. Importantly, this file is not an end product, but a <em>handoff artifact</em>. In the next modules, it will be reused to construct models, build pipelines, and perform inference. Likewise, in the homework, you will repeat this workflow with a different assay and explore an alternative fingerprinting strategy, reinforcing the idea that feature generation is one step in a broader, repeatable process rather than a one-off task. The goal is not to produce a single “correct” dataset, but to understand how each representation fits into a reproducible machine-learning pipeline.</p>
<div class="alert alert-block alert-success">
<H1>Homework</H1>
<strong>Problem 1: hw_10.1a_fp.ipynb</strong> 
<p>Create a new notebook in your notebooks directory called hw_10.1a_fp.ipynb and with this notebook create a new .csv fingerprint file in the <i>/data/AID743139/features</i> directory. You must get instructor approval of the fingerprint and parameters before proceeding and use the modern generator function approach to create the fingerprints.</p>
<p>File naming convention needs to identify</p>
<ul>
    <li>AID</li>
    <li>Finger print type (ecfp, fcfp...)</li>
    <li>Algorithm parameters (radius, size)</li>
        <li>curation decisions</li>
        <li>date and version</li>
</ul>
Example <code>AID743139ecfp4_1024_activites_noSalt_20260111_v1</code> which describes the first version of a 1024 bit extended connectivity fingerprint with a radius of 4 of desalted compounds showing binary activity data for AID743139 that was generated on January 11, 2026.
<br><br>
<strong>Problem 2: hw_10.1_data_prep.ipynb</strong> 
<p>Create a new notebook in your notebooks directory called hw_10.2_data_prep.ipynb and repeat the work of this notebook with the bioassay you choose in your first homework assignment (AID999999 in the directory structure below). You need generate the structure with your notebook the way we did it with this one. 
<pre><code>
10_SupervisedML/
├── data/                    
│   │── AID743139/
│   │ 
│   │── AID999999/   (your AID)   
├── models/                    
│   │── AID743139/
│   │ 
│   │── AID999999/   (your AID)     
</code></pre>
</div></section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "nbi"
        },
        kernelOptions: {
            name: "nbi",
            path: "./content/modules/10_SupervisedML"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'nbi'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="10_0_Intro.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">10.0: Intro</p>
      </div>
    </a>
    <a class="right-next"
       href="../../appendices/App_1/README.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Appendix 1.1: Getting Set Up</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">10.1: Data Preparation and Feature Engineering</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#import-bioactivity-data-from-pubchem">1. Import bioactivity data from PubChem</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#directory-architecture">1.1 Directory Architecture</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#helper-function-make-fixed-filename">1.2 Helper Function: make_fixed_filename</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#data-cleaning">2  Data Cleaning</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#inspect-data-and-remove-first-3-rows">2.1 Inspect data and remove first 3 rows</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#save-versioned-dataframe-as-csv">2.2 Save Versioned DataFrame as CSV</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#helper-function-make-filename">2.2.1 Helper Function: make_filename</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reload-saved-file-into-dataframe">2.2.2 Reload saved file into dataframe</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#inspect-column-names">2.3 Inspect column names</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#clean-column-names">2.3.1 Clean column names</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#inspect-activity-outcomes">3. Inspect Activity Outcomes</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#inspect-activity-summaries">3.1 Inspect Activity summaries</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#select-active-inactive-compounds-for-model-building">4. Select active/inactive compounds for model building</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#select-active-and-inactive-compounds">4.1 Select active and inactive compounds</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#drop-substances-without-associated-cids">4.2 Drop substances without associated CIDs.</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#remove-cids-with-conflicting-activities">4.3 Remove CIDs with conflicting activities</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#remove-redundant-data">4.4 Remove redundant data</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#save-as-aid743139-clean-cid">Save as AID743139_clean_CID…</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#adding-numeric-activity-classes">4.5 Adding “numeric” activity classes</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#create-a-smaller-data-frame-that-only-contains-cids-and-activities">4.6 Create a smaller data frame that only contains CIDs and activities.</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#save-as-aid743139-activity-cid">4.6.1 Save as AID743139_activity_CID</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#acquire-standardized-pubchem-smiles">5 Acquire standardized PubChem SMILES</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#deal-with-salts-and-mixtures">5.1  Deal with salts and mixtures</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#comparing-structures-using-inchi">5.2 Comparing structures using InChI</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#save-as-aid743139-nosalt-pcsmiles">Save as AID743139_noSalt_pcSMILES…</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#convert-smiles-to-binary-data-for-model-input">6 Convert SMILES to Binary Data for Model Input</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-maccs-keys-from-smiles">6.1 Generate MACCS keys from SMILES</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#save-as-features-aid743139-maccs-nosalt">Save as features/AID743139_MACCS_noSalt…</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#merge-activity-data-and-fingerprint-information">6.2 Merge activity data and fingerprint information</a></li>
</ul>
</li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Robert Belford
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright CC 4.0 2026 Robert Belford, additional info on each notebook.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
<div class="extra_footer">
  <script async src="https://hypothes.is/embed.js"></script>

</div>
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>